diff --git a/.vscode/settings.json b/.vscode/settings.json
index d2c0b22..a8d772a 100644
--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -12,6 +12,15 @@
         "string": "c",
         "atomic": "c",
         "thread.h": "c",
-        "vaddr.h": "c"
+        "vaddr.h": "c",
+        "stdbool.h": "c",
+        "process.h": "c",
+        "cstdlib": "c",
+        "init.h": "c",
+        "stdio.h": "c",
+        "mmu.h": "c",
+        "hash.h": "c",
+        "palloc.h": "c",
+        "memory": "c"
     }
 }
\ No newline at end of file
diff --git a/.gitignore b/pintos/.gitignore
similarity index 100%
rename from .gitignore
rename to pintos/.gitignore
diff --git a/pintos/include/threads/thread.h b/pintos/include/threads/thread.h
index 8f172c7..c838a1e 100644
--- a/pintos/include/threads/thread.h
+++ b/pintos/include/threads/thread.h
@@ -5,11 +5,13 @@
 #include <list.h>
 #include <stdint.h>
 #include "threads/interrupt.h"
+#include "threads/synch.h"
 #ifdef VM
 #include "vm/vm.h"
 #endif
 
 
+
 /* States in a thread's life cycle. */
 enum thread_status {
 	THREAD_RUNNING,     /* Running thread. */
@@ -100,6 +102,11 @@ struct thread {
 	/* Owned by userprog/process.c. */
 	uint64_t *pml4;                     /* Page map level 4 */
 	struct fd_table *fd_table;
+	struct thread* parent;
+	struct semaphore self_jail;
+	struct list child_dying_list;
+	struct dying_msg* dmsg;
+
 #endif
 #ifdef VM
 	/* Table for whole virtual memory owned by thread. */
diff --git a/pintos/include/userprog/process.h b/pintos/include/userprog/process.h
index 6ec0a80..f1f7480 100644
--- a/pintos/include/userprog/process.h
+++ b/pintos/include/userprog/process.h
@@ -13,6 +13,13 @@ struct fd_value {
 	struct file *file;
 };
 
+struct dying_msg
+{
+	tid_t tid;
+	int dying_msg;
+	struct list_elem elem;
+};
+
 /* 프로세스마다 파일 디스크립터를 관리하기 위한 구조체 */
 struct fd_table {
 	struct fd_value fd_node[64];
@@ -30,4 +37,6 @@ int process_wait (tid_t);
 void process_exit (void);
 void process_activate (struct thread *next);
 
+struct dying_msg* dying_msg_init (tid_t tid );
+
 #endif /* userprog/process.h */
diff --git a/pintos/include/userprog/syscall.h b/pintos/include/userprog/syscall.h
index d7ac7a3..8b28a4b 100644
--- a/pintos/include/userprog/syscall.h
+++ b/pintos/include/userprog/syscall.h
@@ -2,11 +2,25 @@
 #define USERPROG_SYSCALL_H
 
 #include <stdbool.h>
+#include <stdio.h>
+#include <syscall-nr.h>
+#include "threads/interrupt.h"
+#include "threads/thread.h"
+#include "threads/loader.h"
+#include "userprog/gdt.h"
+#include "threads/flags.h"
+#include "intrinsic.h"
+#include "filesys/file.h"
+#include "filesys/filesys.h"
+#include "userprog/process.h"
+
 
 void syscall_init (void);
 
 void sys_halt (void);
 void sys_exit (int status);
+int sys_wait (tid_t tid);
+tid_t sys_fork (const char *thread_name,struct intr_frame * _if);
 bool sys_create (const char *file, unsigned initial_size);
 int sys_open (const char *file);
 int sys_filesize (int fd);
diff --git a/pintos/threads/thread.c b/pintos/threads/thread.c
index 020af08..45e2790 100644
--- a/pintos/threads/thread.c
+++ b/pintos/threads/thread.c
@@ -11,6 +11,7 @@
 #include "threads/synch.h"
 #include "threads/vaddr.h"
 #include "intrinsic.h"
+#include "threads/malloc.h"
 #ifdef USERPROG
 #include "userprog/process.h"
 #endif
@@ -44,27 +45,27 @@ static struct lock tid_lock;
 static struct list destruction_req;
 
 /* Statistics. */
-static long long idle_ticks;    /* # of timer ticks spent idle. */
-static long long kernel_ticks;  /* # of timer ticks in kernel threads. */
-static long long user_ticks;    /* # of timer ticks in user programs. */
+static long long idle_ticks;   /* # of timer ticks spent idle. */
+static long long kernel_ticks; /* # of timer ticks in kernel threads. */
+static long long user_ticks;   /* # of timer ticks in user programs. */
 
 /* Scheduling. */
-#define TIME_SLICE 4            /* # of timer ticks to give each thread. */
-static unsigned thread_ticks;   /* # of timer ticks since last yield. */
+#define TIME_SLICE 4		  /* # of timer ticks to give each thread. */
+static unsigned thread_ticks; /* # of timer ticks since last yield. */
 
 /* If false (default), use round-robin scheduler.
    If true, use multi-level feedback queue scheduler.
    Controlled by kernel command-line option "-o mlfqs". */
 bool thread_mlfqs;
 
-static void kernel_thread (thread_func *, void *aux);
+static void kernel_thread(thread_func *, void *aux);
 
-static void idle (void *aux UNUSED);
-static struct thread *next_thread_to_run (void);
-static void init_thread (struct thread *, const char *name, int priority);
+static void idle(void *aux UNUSED);
+static struct thread *next_thread_to_run(void);
+static void init_thread(struct thread *, const char *name, int priority);
 static void do_schedule(int status);
-static void schedule (void);
-static tid_t allocate_tid (void);
+static void schedule(void);
+static tid_t allocate_tid(void);
 
 /* Returns true if T appears to point to a valid thread. */
 #define is_thread(t) ((t) != NULL && (t)->magic == THREAD_MAGIC)
@@ -74,36 +75,35 @@ static tid_t allocate_tid (void);
  * down to the start of a page.  Since `struct thread' is
  * always at the beginning of a page and the stack pointer is
  * somewhere in the middle, this locates the curent thread. */
-#define running_thread() ((struct thread *) (pg_round_down (rrsp ())))
-
+#define running_thread() ((struct thread *)(pg_round_down(rrsp())))
 
 // Global descriptor table for the thread_start.
 // Because the gdt will be setup after the thread_init, we should
 // setup temporal gdt first.
-static uint64_t gdt[3] = { 0, 0x00af9a000000ffff, 0x00cf92000000ffff };
+static uint64_t gdt[3] = {0, 0x00af9a000000ffff, 0x00cf92000000ffff};
 
 /* list_insert_ordered 함수에서 사용하기 위한 함수
 	priority를 높은 순으로 삽입하기 위한 함수 */
 static bool
-prior_priority (const struct list_elem *a_, const struct list_elem *b_,
-            void *aux UNUSED) 
+prior_priority(const struct list_elem *a_, const struct list_elem *b_,
+			   void *aux UNUSED)
 {
-  const struct thread *a = list_entry (a_, struct thread, elem);
-  const struct thread *b = list_entry (b_, struct thread, elem);
-  
-  return a->priority > b->priority;
+	const struct thread *a = list_entry(a_, struct thread, elem);
+	const struct thread *b = list_entry(b_, struct thread, elem);
+
+	return a->priority > b->priority;
 }
 
 /* list_insert_ordered 함수에서 사용하기 위한 함수
 	awake_ticks를 낮은 순으로 삽입하기 위한 함수 */
 static bool
-prior_tick (const struct list_elem *a_, const struct list_elem *b_,
-            void *aux UNUSED) 
+prior_tick(const struct list_elem *a_, const struct list_elem *b_,
+		   void *aux UNUSED)
 {
-  const struct thread *a = list_entry (a_, struct thread, elem);
-  const struct thread *b = list_entry (b_, struct thread, elem);
-  
-  return a->awake_ticks < b->awake_ticks;
+	const struct thread *a = list_entry(a_, struct thread, elem);
+	const struct thread *b = list_entry(b_, struct thread, elem);
+
+	return a->awake_ticks < b->awake_ticks;
 }
 
 /* Initializes the threading system by transforming the code
@@ -118,55 +118,54 @@ prior_tick (const struct list_elem *a_, const struct list_elem *b_,
    thread_create().
 
    It is not safe to call thread_current() until this function
-   finishes. 
+   finishes.
    pintos 운영체제에서 커널 스레딩 시스템을 초기화 할때 사용하는 함수로, os main함수에서 최초 한번 동작 */
-void
-thread_init (void) {
-	ASSERT (intr_get_level () == INTR_OFF);
+void thread_init(void)
+{
+	ASSERT(intr_get_level() == INTR_OFF);
 
 	/* Reload the temporal gdt for the kernel
 	 * This gdt does not include the user context.
 	 * The kernel will rebuild the gdt with user context, in gdt_init (). */
 	struct desc_ptr gdt_ds = {
-		.size = sizeof (gdt) - 1,
-		.address = (uint64_t) gdt
-	};
-	lgdt (&gdt_ds);
+		.size = sizeof(gdt) - 1,
+		.address = (uint64_t)gdt};
+	lgdt(&gdt_ds);
 
 	/* Init the globla thread context */
-	lock_init (&tid_lock);
-	list_init (&ready_list);
-	list_init (&sleep_list);
-	list_init (&destruction_req);
-
+	lock_init(&tid_lock);
+	list_init(&ready_list);
+	list_init(&sleep_list);
+	list_init(&destruction_req);
+	
 	/* Set up a thread structure for the running thread. */
-	initial_thread = running_thread ();
-	init_thread (initial_thread, "main", PRI_DEFAULT);
+	initial_thread = running_thread();
+	init_thread(initial_thread, "main", PRI_DEFAULT);
 	initial_thread->status = THREAD_RUNNING;
-	initial_thread->tid = allocate_tid ();
+	initial_thread->tid = allocate_tid();
 }
 
 /* Starts preemptive thread scheduling by enabling interrupts.
    Also creates the idle thread. */
-void
-thread_start (void) {
+void thread_start(void)
+{
 	/* Create the idle thread. */
 	struct semaphore idle_started;
-	sema_init (&idle_started, 0);
-	thread_create ("idle", PRI_MIN, idle, &idle_started);
+	sema_init(&idle_started, 0);
+	thread_create("idle", PRI_MIN, idle, &idle_started);
 
 	/* Start preemptive thread scheduling. */
-	intr_enable ();
+	intr_enable();
 
 	/* Wait for the idle thread to initialize idle_thread. */
-	sema_down (&idle_started);
+	sema_down(&idle_started);
 }
 
 /* Called by the timer interrupt handler at each timer tick.
    Thus, this function runs in an external interrupt context. */
-void
-thread_tick (void) {
-	struct thread *t = thread_current ();
+void thread_tick(void)
+{
+	struct thread *t = thread_current();
 
 	/* Update statistics. */
 	if (t == idle_thread)
@@ -180,14 +179,14 @@ thread_tick (void) {
 
 	/* Enforce preemption. */
 	if (++thread_ticks >= TIME_SLICE)
-		intr_yield_on_return ();
+		intr_yield_on_return();
 }
 
 /* Prints thread statistics. */
-void
-thread_print_stats (void) {
-	printf ("Thread: %lld idle ticks, %lld kernel ticks, %lld user ticks\n",
-			idle_ticks, kernel_ticks, user_ticks);
+void thread_print_stats(void)
+{
+	printf("Thread: %lld idle ticks, %lld kernel ticks, %lld user ticks\n",
+		   idle_ticks, kernel_ticks, user_ticks);
 }
 
 /* Creates a new kernel thread named NAME with the given initial
@@ -205,37 +204,40 @@ thread_print_stats (void) {
    The code provided sets the new thread's `priority' member to
    PRIORITY, but no actual priority scheduling is implemented.
    Priority scheduling is the goal of Problem 1-3. */
-tid_t
-thread_create (const char *name, int priority,
-		thread_func *function, void *aux) {
+tid_t thread_create(const char *name, int priority,
+					thread_func *function, void *aux)
+{
 	struct thread *t;
 	tid_t tid;
 
-	ASSERT (function != NULL);
+	ASSERT(function != NULL);
 
 	/* Allocate thread. */
-	t = palloc_get_page (PAL_ZERO);
+	t = palloc_get_page(PAL_ZERO);
 	if (t == NULL)
 		return TID_ERROR;
 
 	/* Initialize thread. */
-	init_thread (t, name, priority);
-	tid = t->tid = allocate_tid ();
+	init_thread(t, name, priority);
+	tid = t->tid = allocate_tid();
 
 	/* Call the kernel_thread if it scheduled.
 	 * Note) rdi is 1st argument, and rsi is 2nd argument. */
-	t->tf.rip = (uintptr_t) kernel_thread;
-	t->tf.R.rdi = (uint64_t) function;
-	t->tf.R.rsi = (uint64_t) aux;
+	t->tf.rip = (uintptr_t)kernel_thread;
+	t->tf.R.rdi = (uint64_t)function;
+	t->tf.R.rsi = (uint64_t)aux;
 	t->tf.ds = SEL_KDSEG;
 	t->tf.es = SEL_KDSEG;
 	t->tf.ss = SEL_KDSEG;
 	t->tf.cs = SEL_KCSEG;
 	t->tf.eflags = FLAG_IF;
 
+	t->dmsg = malloc(sizeof(struct dying_msg));
+	t->dmsg->tid = t->tid;
+
 	/* Add to run queue. */
-	thread_unblock (t);
-	thread_maybe_yield();	// 나보다 우선순위가 큰 스레드가 생성될 수 있으니까 체크
+	thread_unblock(t);
+	thread_maybe_yield(); // 나보다 우선순위가 큰 스레드가 생성될 수 있으니까 체크
 
 	return tid;
 }
@@ -246,12 +248,12 @@ thread_create (const char *name, int priority,
    This function must be called with interrupts turned off.  It
    is usually a better idea to use one of the synchronization
    primitives in synch.h. */
-void
-thread_block (void) {
-	ASSERT (!intr_context ());
-	ASSERT (intr_get_level () == INTR_OFF);
-	thread_current ()->status = THREAD_BLOCKED;
-	schedule ();
+void thread_block(void)
+{
+	ASSERT(!intr_context());
+	ASSERT(intr_get_level() == INTR_OFF);
+	thread_current()->status = THREAD_BLOCKED;
+	schedule();
 }
 
 /* Transitions a blocked thread T to the ready-to-run state.
@@ -262,183 +264,189 @@ thread_block (void) {
    be important: if the caller had disabled interrupts itself,
    it may expect that it can atomically unblock a thread and
    update other data. */
-void
-thread_unblock (struct thread *t) {
+void thread_unblock(struct thread *t)
+{
 	enum intr_level old_level;
 
-	ASSERT (is_thread (t));
+	ASSERT(is_thread(t));
 
-	old_level = intr_disable ();
-	ASSERT (t->status == THREAD_BLOCKED);
-	list_insert_ordered (&ready_list, &t->elem, prior_priority, NULL);	// 삽입도 priority 순위에 맞게 삽입
+	old_level = intr_disable();
+	ASSERT(t->status == THREAD_BLOCKED);
+	list_insert_ordered(&ready_list, &t->elem, prior_priority, NULL); // 삽입도 priority 순위에 맞게 삽입
 	t->status = THREAD_READY;
-	intr_set_level (old_level);
+	intr_set_level(old_level);
 }
 
 /* 현재 스레드를 sleep 하는 함수
 	이 함수를 사용하는 순간의 스레드를 sleep_list에 넣는다
 	(이미 running 상태여서 ready_list에서 pop 되었기 떄문에 따로 list_remove는 필요없음) */
-void
-thread_sleep (int64_t alarm_ticks)
+void thread_sleep(int64_t alarm_ticks)
 {
 	enum intr_level old_level;
 	struct thread *current_thread;
 
-	old_level = intr_disable ();
+	old_level = intr_disable();
 
-	current_thread = thread_current ();
+	current_thread = thread_current();
 	ASSERT(current_thread->status == THREAD_RUNNING);
 	current_thread->awake_ticks = alarm_ticks;
-	list_insert_ordered (&sleep_list, &current_thread->elem, prior_tick, NULL);
-	thread_block ();
+	list_insert_ordered(&sleep_list, &current_thread->elem, prior_tick, NULL);
+	thread_block();
 
-	intr_set_level (old_level);
+	intr_set_level(old_level);
 }
 
 /* 자고있는 스레드들을 깨우는 함수
 	매개변수로 현재 tick을 받으며, sleep_list의 가장 앞의 스레드의
 	awake_ticks과 비교하여 현재 tick이 넘어갈 경우 해당 스레드를 unblock 한다 */
-void
-thread_wakeup (int64_t current_ticks)
+void thread_wakeup(int64_t current_ticks)
 {
 	struct thread *current_thread;
 	enum intr_level old_level;
 
-	if (!list_empty (&sleep_list)) {
-		old_level = intr_disable ();
+	if (!list_empty(&sleep_list))
+	{
+		old_level = intr_disable();
 
-		for (struct list_elem *elem = list_begin(&sleep_list); elem != list_end(&sleep_list); elem = list_begin(&sleep_list)) {
+		for (struct list_elem *elem = list_begin(&sleep_list); elem != list_end(&sleep_list); elem = list_begin(&sleep_list))
+		{
 			current_thread = list_entry(elem, struct thread, elem);
 
-			if (current_thread->awake_ticks <= current_ticks) {
+			if (current_thread->awake_ticks <= current_ticks)
+			{
 				list_remove(elem);
 				thread_unblock(current_thread);
 			}
-			else {
+			else
+			{
 				break;
 			}
 		}
 
-		intr_set_level (old_level);
+		intr_set_level(old_level);
 	}
 }
 
 /* Returns the name of the running thread. */
 const char *
-thread_name (void) {
-	return thread_current ()->name;
+thread_name(void)
+{
+	return thread_current()->name;
 }
 
 /* Returns the running thread.
    This is running_thread() plus a couple of sanity checks.
    See the big comment at the top of thread.h for details. */
 struct thread *
-thread_current (void) {
-	struct thread *t = running_thread ();
+thread_current(void)
+{
+	struct thread *t = running_thread();
 
 	/* Make sure T is really a thread.
 	   If either of these assertions fire, then your thread may
 	   have overflowed its stack.  Each thread has less than 4 kB
 	   of stack, so a few big automatic arrays or moderate
 	   recursion can cause stack overflow. */
-	ASSERT (is_thread (t));
-	ASSERT (t->status == THREAD_RUNNING);
+	ASSERT(is_thread(t));
+	ASSERT(t->status == THREAD_RUNNING);
 
 	return t;
 }
 
 /* Returns the running thread's tid. */
-tid_t
-thread_tid (void) {
-	return thread_current ()->tid;
+tid_t thread_tid(void)
+{
+	return thread_current()->tid;
 }
 
 /* Deschedules the current thread and destroys it.  Never
    returns to the caller. */
-void
-thread_exit (void) {
-	ASSERT (!intr_context ());
+void thread_exit(void)
+{
+	ASSERT(!intr_context());
 
 #ifdef USERPROG
-	process_exit ();
+	if(thread_current()->parent != NULL){process_exit();}
+	
 	// pricess.c 안에서 모든 fd를 close 하고 이후에 palloc한거 제거하는거 구현
 #endif
 
 	/* Just set our status to dying and schedule another process.
 	   We will be destroyed during the call to schedule_tail(). */
-	intr_disable ();
-	do_schedule (THREAD_DYING);
-	NOT_REACHED ();
+	intr_disable();
+	do_schedule(THREAD_DYING);
+	NOT_REACHED();
 }
 
 /* Yields the CPU.  The current thread is not put to sleep and
-   may be scheduled again immediately at the scheduler's whim. 
+   may be scheduled again immediately at the scheduler's whim.
    현재 스레드를 read 상태로 만든 뒤 ready_list에 넣고 스케쥴링 동작하는 함수 */
-void
-thread_yield (void) {
-	struct thread *curr = thread_current ();
+void thread_yield(void)
+{
+	struct thread *curr = thread_current();
 	enum intr_level old_level;
 
-	ASSERT (!intr_context ());
+	ASSERT(!intr_context());
 
-	old_level = intr_disable ();
+	old_level = intr_disable();
 	if (curr != idle_thread)
-		list_insert_ordered (&ready_list, &curr->elem, prior_priority, NULL);	// 삽입도 priority 순위에 맞게 삽입
-	do_schedule (THREAD_READY);
-	intr_set_level (old_level);
+		list_insert_ordered(&ready_list, &curr->elem, prior_priority, NULL); // 삽입도 priority 순위에 맞게 삽입
+	do_schedule(THREAD_READY);
+	intr_set_level(old_level);
 }
 
 /* 현재 스레드가 ready_list의 앞의 스레드와 비교하고 스위칭 하는 함수  */
-void
-thread_maybe_yield (void) {
-	enum intr_level old_level = intr_disable ();
+void thread_maybe_yield(void)
+{
+	enum intr_level old_level = intr_disable();
 
-	if (!list_empty (&ready_list) && thread_current ()->priority < list_entry (list_front (&ready_list), struct thread, elem)->priority) {
+	if (!list_empty(&ready_list) && thread_current()->priority < list_entry(list_front(&ready_list), struct thread, elem)->priority)
+	{
 		if (intr_context())
 			intr_yield_on_return();
-        else
+		else
 			thread_yield();
 	}
 
-	intr_set_level (old_level);
+	intr_set_level(old_level);
 }
 
 /* Sets the current thread's priority to NEW_PRIORITY. */
-void
-thread_set_priority (int new_priority) {
-	thread_current ()->priority = new_priority;
-	thread_maybe_yield();	// 세팅된 priority가 낮은 순위일 수 있으니까 체크
+void thread_set_priority(int new_priority)
+{
+	thread_current()->priority = new_priority;
+	thread_maybe_yield(); // 세팅된 priority가 낮은 순위일 수 있으니까 체크
 }
 
 /* Returns the current thread's priority. */
-int
-thread_get_priority (void) {
-	return thread_current ()->priority;
+int thread_get_priority(void)
+{
+	return thread_current()->priority;
 }
 
 /* Sets the current thread's nice value to NICE. */
-void
-thread_set_nice (int nice UNUSED) {
+void thread_set_nice(int nice UNUSED)
+{
 	/* TODO: Your implementation goes here */
 }
 
 /* Returns the current thread's nice value. */
-int
-thread_get_nice (void) {
+int thread_get_nice(void)
+{
 	/* TODO: Your implementation goes here */
 	return 0;
 }
 
 /* Returns 100 times the system load average. */
-int
-thread_get_load_avg (void) {
+int thread_get_load_avg(void)
+{
 	/* TODO: Your implementation goes here */
 	return 0;
 }
 
 /* Returns 100 times the current thread's recent_cpu value. */
-int
-thread_get_recent_cpu (void) {
+int thread_get_recent_cpu(void)
+{
 	/* TODO: Your implementation goes here */
 	return 0;
 }
@@ -453,16 +461,18 @@ thread_get_recent_cpu (void) {
    ready list.  It is returned by next_thread_to_run() as a
    special case when the ready list is empty. */
 static void
-idle (void *idle_started_ UNUSED) {
+idle(void *idle_started_ UNUSED)
+{
 	struct semaphore *idle_started = idle_started_;
 
-	idle_thread = thread_current ();
-	sema_up (idle_started);
+	idle_thread = thread_current();
+	sema_up(idle_started);
 
-	for (;;) {
+	for (;;)
+	{
 		/* Let someone else run. */
-		intr_disable ();
-		thread_block ();
+		intr_disable();
+		thread_block();
 
 		/* Re-enable interrupts and wait for the next one.
 
@@ -476,77 +486,88 @@ idle (void *idle_started_ UNUSED) {
 
 		   See [IA32-v2a] "HLT", [IA32-v2b] "STI", and [IA32-v3a]
 		   7.11.1 "HLT Instruction". */
-		asm volatile ("sti; hlt" : : : "memory");
+		asm volatile("sti; hlt" : : : "memory");
 	}
 }
 
 /* Function used as the basis for a kernel thread. */
 static void
-kernel_thread (thread_func *function, void *aux) {
-	ASSERT (function != NULL);
+kernel_thread(thread_func *function, void *aux)
+{
+	ASSERT(function != NULL);
 
-	intr_enable ();       /* The scheduler runs with interrupts off. */
-	function (aux);       /* Execute the thread function. */
-	thread_exit ();       /* If function() returns, kill the thread. */
+	intr_enable(); /* The scheduler runs with interrupts off. */
+	function(aux); /* Execute the thread function. */
+	thread_exit(); /* If function() returns, kill the thread. */
 }
 
-
 /* Does basic initialization of T as a blocked thread named
    NAME. */
 static void
-init_thread (struct thread *t, const char *name, int priority) {
-	ASSERT (t != NULL);
-	ASSERT (PRI_MIN <= priority && priority <= PRI_MAX);
-	ASSERT (name != NULL);
+init_thread(struct thread *t, const char *name, int priority)
+{
+	ASSERT(t != NULL);
+	ASSERT(PRI_MIN <= priority && priority <= PRI_MAX);
+	ASSERT(name != NULL);
 
-	memset (t, 0, sizeof *t);
+	memset(t, 0, sizeof *t);
 	t->status = THREAD_BLOCKED;
-	strlcpy (t->name, name, sizeof t->name);
-	t->tf.rsp = (uint64_t) t + PGSIZE - sizeof (void *);
+	strlcpy(t->name, name, sizeof t->name);
+	t->tf.rsp = (uint64_t)t + PGSIZE - sizeof(void *);
 	t->priority = priority;
 	t->magic = THREAD_MAGIC;
+
+	#ifdef USERPROG
+	list_init(&t->child_dying_list);
+	list_init(&t->self_jail.waiters);
+	sema_init(&t->self_jail,0);
+	// t->dmsg = dying_msg_init(t->tid);
+
+	#endif
+	
 }
 
 /* Chooses and returns the next thread to be scheduled.  Should
    return a thread from the run queue, unless the run queue is
    empty.  (If the running thread can continue running, then it
    will be in the run queue.)  If the run queue is empty, return
-   idle_thread. 
+   idle_thread.
    ready_list에서 다음 실행할 스레드를 선택해서 반환, 만약 리스트에 아무것도 없을 경우 idle_thread를 반환  */
 static struct thread *
-next_thread_to_run (void) {
-	if (list_empty (&ready_list))
+next_thread_to_run(void)
+{
+	if (list_empty(&ready_list))
 		return idle_thread;
 	else
-		return list_entry (list_pop_front (&ready_list), struct thread, elem);
+		return list_entry(list_pop_front(&ready_list), struct thread, elem);
 }
 
 /* Use iretq to launch the thread */
-void
-do_iret (struct intr_frame *tf) {
+void do_iret(struct intr_frame *tf)
+{
 	__asm __volatile(
-			"movq %0, %%rsp\n"
-			"movq 0(%%rsp),%%r15\n"
-			"movq 8(%%rsp),%%r14\n"
-			"movq 16(%%rsp),%%r13\n"
-			"movq 24(%%rsp),%%r12\n"
-			"movq 32(%%rsp),%%r11\n"
-			"movq 40(%%rsp),%%r10\n"
-			"movq 48(%%rsp),%%r9\n"
-			"movq 56(%%rsp),%%r8\n"
-			"movq 64(%%rsp),%%rsi\n"
-			"movq 72(%%rsp),%%rdi\n"
-			"movq 80(%%rsp),%%rbp\n"
-			"movq 88(%%rsp),%%rdx\n"
-			"movq 96(%%rsp),%%rcx\n"
-			"movq 104(%%rsp),%%rbx\n"
-			"movq 112(%%rsp),%%rax\n"
-			"addq $120,%%rsp\n"
-			"movw 8(%%rsp),%%ds\n"
-			"movw (%%rsp),%%es\n"
-			"addq $32, %%rsp\n"
-			"iretq"
-			: : "g" ((uint64_t) tf) : "memory");
+		"movq %0, %%rsp\n"
+		"movq 0(%%rsp),%%r15\n"
+		"movq 8(%%rsp),%%r14\n"
+		"movq 16(%%rsp),%%r13\n"
+		"movq 24(%%rsp),%%r12\n"
+		"movq 32(%%rsp),%%r11\n"
+		"movq 40(%%rsp),%%r10\n"
+		"movq 48(%%rsp),%%r9\n"
+		"movq 56(%%rsp),%%r8\n"
+		"movq 64(%%rsp),%%rsi\n"
+		"movq 72(%%rsp),%%rdi\n"
+		"movq 80(%%rsp),%%rbp\n"
+		"movq 88(%%rsp),%%rdx\n"
+		"movq 96(%%rsp),%%rcx\n"
+		"movq 104(%%rsp),%%rbx\n"
+		"movq 112(%%rsp),%%rax\n"
+		"addq $120,%%rsp\n"
+		"movw 8(%%rsp),%%ds\n"
+		"movw (%%rsp),%%es\n"
+		"addq $32, %%rsp\n"
+		"iretq"
+		: : "g"((uint64_t)tf) : "memory");
 }
 
 /* Switching the thread by activating the new thread's page
@@ -560,62 +581,62 @@ do_iret (struct intr_frame *tf) {
    complete.  In practice that means that printf()s should be
    added at the end of the function. */
 static void
-thread_launch (struct thread *th) {
-	uint64_t tf_cur = (uint64_t) &running_thread ()->tf;
-	uint64_t tf = (uint64_t) &th->tf;
-	ASSERT (intr_get_level () == INTR_OFF);
+thread_launch(struct thread *th)
+{
+	uint64_t tf_cur = (uint64_t)&running_thread()->tf;
+	uint64_t tf = (uint64_t)&th->tf;
+	ASSERT(intr_get_level() == INTR_OFF);
 
 	/* The main switching logic.
 	 * We first restore the whole execution context into the intr_frame
 	 * and then switching to the next thread by calling do_iret.
 	 * Note that, we SHOULD NOT use any stack from here
 	 * until switching is done. */
-	__asm __volatile (
-			/* Store registers that will be used. */
-			"push %%rax\n"
-			"push %%rbx\n"
-			"push %%rcx\n"
-			/* Fetch input once */
-			"movq %0, %%rax\n"
-			"movq %1, %%rcx\n"
-			"movq %%r15, 0(%%rax)\n"
-			"movq %%r14, 8(%%rax)\n"
-			"movq %%r13, 16(%%rax)\n"
-			"movq %%r12, 24(%%rax)\n"
-			"movq %%r11, 32(%%rax)\n"
-			"movq %%r10, 40(%%rax)\n"
-			"movq %%r9, 48(%%rax)\n"
-			"movq %%r8, 56(%%rax)\n"
-			"movq %%rsi, 64(%%rax)\n"
-			"movq %%rdi, 72(%%rax)\n"
-			"movq %%rbp, 80(%%rax)\n"
-			"movq %%rdx, 88(%%rax)\n"
-			"pop %%rbx\n"              // Saved rcx
-			"movq %%rbx, 96(%%rax)\n"
-			"pop %%rbx\n"              // Saved rbx
-			"movq %%rbx, 104(%%rax)\n"
-			"pop %%rbx\n"              // Saved rax
-			"movq %%rbx, 112(%%rax)\n"
-			"addq $120, %%rax\n"
-			"movw %%es, (%%rax)\n"
-			"movw %%ds, 8(%%rax)\n"
-			"addq $32, %%rax\n"
-			"call __next\n"         // read the current rip.
-			"__next:\n"
-			"pop %%rbx\n"
-			"addq $(out_iret -  __next), %%rbx\n"
-			"movq %%rbx, 0(%%rax)\n" // rip
-			"movw %%cs, 8(%%rax)\n"  // cs
-			"pushfq\n"
-			"popq %%rbx\n"
-			"mov %%rbx, 16(%%rax)\n" // eflags
-			"mov %%rsp, 24(%%rax)\n" // rsp
-			"movw %%ss, 32(%%rax)\n"
-			"mov %%rcx, %%rdi\n"
-			"call do_iret\n"
-			"out_iret:\n"
-			: : "g"(tf_cur), "g" (tf) : "memory"
-			);
+	__asm __volatile(
+		/* Store registers that will be used. */
+		"push %%rax\n"
+		"push %%rbx\n"
+		"push %%rcx\n"
+		/* Fetch input once */
+		"movq %0, %%rax\n"
+		"movq %1, %%rcx\n"
+		"movq %%r15, 0(%%rax)\n"
+		"movq %%r14, 8(%%rax)\n"
+		"movq %%r13, 16(%%rax)\n"
+		"movq %%r12, 24(%%rax)\n"
+		"movq %%r11, 32(%%rax)\n"
+		"movq %%r10, 40(%%rax)\n"
+		"movq %%r9, 48(%%rax)\n"
+		"movq %%r8, 56(%%rax)\n"
+		"movq %%rsi, 64(%%rax)\n"
+		"movq %%rdi, 72(%%rax)\n"
+		"movq %%rbp, 80(%%rax)\n"
+		"movq %%rdx, 88(%%rax)\n"
+		"pop %%rbx\n" // Saved rcx
+		"movq %%rbx, 96(%%rax)\n"
+		"pop %%rbx\n" // Saved rbx
+		"movq %%rbx, 104(%%rax)\n"
+		"pop %%rbx\n" // Saved rax
+		"movq %%rbx, 112(%%rax)\n"
+		"addq $120, %%rax\n"
+		"movw %%es, (%%rax)\n"
+		"movw %%ds, 8(%%rax)\n"
+		"addq $32, %%rax\n"
+		"call __next\n" // read the current rip.
+		"__next:\n"
+		"pop %%rbx\n"
+		"addq $(out_iret -  __next), %%rbx\n"
+		"movq %%rbx, 0(%%rax)\n" // rip
+		"movw %%cs, 8(%%rax)\n"	 // cs
+		"pushfq\n"
+		"popq %%rbx\n"
+		"mov %%rbx, 16(%%rax)\n" // eflags
+		"mov %%rsp, 24(%%rax)\n" // rsp
+		"movw %%ss, 32(%%rax)\n"
+		"mov %%rcx, %%rdi\n"
+		"call do_iret\n"
+		"out_iret:\n"
+		: : "g"(tf_cur), "g"(tf) : "memory");
 }
 
 /* Schedules a new process. At entry, interrupts must be off.
@@ -624,26 +645,29 @@ thread_launch (struct thread *th) {
  * It's not safe to call printf() in the schedule().
  * 현재 실행중인 스레드를 해당 status로 바꾸고 다음에 실행할 스레드를 선택하여 스케줄링 하는 함수  */
 static void
-do_schedule(int status) {
-	ASSERT (intr_get_level () == INTR_OFF);
-	ASSERT (thread_current()->status == THREAD_RUNNING);
-	while (!list_empty (&destruction_req)) {
+do_schedule(int status)
+{
+	ASSERT(intr_get_level() == INTR_OFF);
+	ASSERT(thread_current()->status == THREAD_RUNNING);
+	while (!list_empty(&destruction_req))
+	{
 		struct thread *victim =
-			list_entry (list_pop_front (&destruction_req), struct thread, elem);
+			list_entry(list_pop_front(&destruction_req), struct thread, elem);
 		palloc_free_page(victim);
 	}
-	thread_current ()->status = status;
-	schedule ();
+	thread_current()->status = status;
+	schedule();
 }
 
 static void
-schedule (void) {
-	struct thread *curr = running_thread ();
-	struct thread *next = next_thread_to_run ();
+schedule(void)
+{
+	struct thread *curr = running_thread();
+	struct thread *next = next_thread_to_run();
 
-	ASSERT (intr_get_level () == INTR_OFF);
-	ASSERT (curr->status != THREAD_RUNNING);
-	ASSERT (is_thread (next));
+	ASSERT(intr_get_level() == INTR_OFF);
+	ASSERT(curr->status != THREAD_RUNNING);
+	ASSERT(is_thread(next));
 	/* Mark us as running. */
 	next->status = THREAD_RUNNING;
 
@@ -652,10 +676,11 @@ schedule (void) {
 
 #ifdef USERPROG
 	/* Activate the new address space. */
-	process_activate (next);
+	process_activate(next);
 #endif
 
-	if (curr != next) {
+	if (curr != next)
+	{
 		/* If the thread we switched from is dying, destroy its struct
 		   thread. This must happen late so that thread_exit() doesn't
 		   pull out the rug under itself.
@@ -663,26 +688,28 @@ schedule (void) {
 		   currently used by the stack.
 		   The real destruction logic will be called at the beginning of the
 		   schedule(). */
-		if (curr && curr->status == THREAD_DYING && curr != initial_thread) {
-			ASSERT (curr != next);
-			list_push_back (&destruction_req, &curr->elem);
+		if (curr && curr->status == THREAD_DYING && curr != initial_thread)
+		{
+			ASSERT(curr != next);
+			list_push_back(&destruction_req, &curr->elem);
 		}
 
 		/* Before switching the thread, we first save the information
 		 * of current running. */
-		thread_launch (next);
+		thread_launch(next);
 	}
 }
 
 /* Returns a tid to use for a new thread. */
 static tid_t
-allocate_tid (void) {
+allocate_tid(void)
+{
 	static tid_t next_tid = 1;
 	tid_t tid;
 
-	lock_acquire (&tid_lock);
+	lock_acquire(&tid_lock);
 	tid = next_tid++;
-	lock_release (&tid_lock);
+	lock_release(&tid_lock);
 
 	return tid;
 }
diff --git a/pintos/userprog/.test_status b/pintos/userprog/.test_status
index 7199a29..7316124 100644
--- a/pintos/userprog/.test_status
+++ b/pintos/userprog/.test_status
@@ -21,6 +21,7 @@ open-boundary PASS
 create-null PASS
 open-bad-ptr PASS
 write-normal PASS
+fork-once PASS
 write-boundary PASS
 close-bad-fd PASS
 exit PASS
diff --git a/pintos/userprog/build/devices/intq.o b/pintos/userprog/build/devices/intq.o
index 4d9b476..2f10582 100644
Binary files a/pintos/userprog/build/devices/intq.o and b/pintos/userprog/build/devices/intq.o differ
diff --git a/pintos/userprog/build/devices/serial.o b/pintos/userprog/build/devices/serial.o
index b8e6807..9007782 100644
Binary files a/pintos/userprog/build/devices/serial.o and b/pintos/userprog/build/devices/serial.o differ
diff --git a/pintos/userprog/build/devices/timer.o b/pintos/userprog/build/devices/timer.o
index 6c21d85..b609e31 100644
Binary files a/pintos/userprog/build/devices/timer.o and b/pintos/userprog/build/devices/timer.o differ
diff --git a/pintos/userprog/build/filesys/page_cache.d b/pintos/userprog/build/filesys/page_cache.d
index d82abef..00a80c3 100644
--- a/pintos/userprog/build/filesys/page_cache.d
+++ b/pintos/userprog/build/filesys/page_cache.d
@@ -5,4 +5,4 @@ filesys/page_cache.o: ../../filesys/page_cache.c ../../include/vm/vm.h \
  ../../include/vm/file.h ../../include/filesys/file.h \
  ../../include/filesys/off_t.h ../../include/threads/thread.h \
  ../../include/lib/debug.h ../../include/lib/kernel/list.h \
- ../../include/threads/interrupt.h
+ ../../include/threads/interrupt.h ../../include/threads/synch.h
diff --git a/pintos/userprog/build/tests/threads/alarm-priority.o b/pintos/userprog/build/tests/threads/alarm-priority.o
index c7b2ae0..26a8c00 100644
Binary files a/pintos/userprog/build/tests/threads/alarm-priority.o and b/pintos/userprog/build/tests/threads/alarm-priority.o differ
diff --git a/pintos/userprog/build/tests/threads/alarm-simultaneous.o b/pintos/userprog/build/tests/threads/alarm-simultaneous.o
index 53afe5c..0986c4a 100644
Binary files a/pintos/userprog/build/tests/threads/alarm-simultaneous.o and b/pintos/userprog/build/tests/threads/alarm-simultaneous.o differ
diff --git a/pintos/userprog/build/tests/threads/alarm-wait.o b/pintos/userprog/build/tests/threads/alarm-wait.o
index 7a1404d..1564c3d 100644
Binary files a/pintos/userprog/build/tests/threads/alarm-wait.o and b/pintos/userprog/build/tests/threads/alarm-wait.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-block.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-block.o
index bd0beaa..4e4dee3 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-block.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-block.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-fair.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-fair.o
index f176578..5082f6b 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-fair.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-fair.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-1.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-1.o
index 4064569..3508eab 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-1.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-1.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-60.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-60.o
index 33bc2ad..999236e 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-60.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-60.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-avg.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-avg.o
index 8e7e466..3e194d7 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-avg.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-load-avg.o differ
diff --git a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-recent-1.o b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-recent-1.o
index f624ae6..5a20f5d 100644
Binary files a/pintos/userprog/build/tests/threads/mlfqs/mlfqs-recent-1.o and b/pintos/userprog/build/tests/threads/mlfqs/mlfqs-recent-1.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-change.d b/pintos/userprog/build/tests/threads/priority-change.d
index 92dcb40..23dfa95 100644
--- a/pintos/userprog/build/tests/threads/priority-change.d
+++ b/pintos/userprog/build/tests/threads/priority-change.d
@@ -4,4 +4,5 @@ tests/threads/priority-change.o: ../../tests/threads/priority-change.c \
  ../../include/lib/stddef.h ../../include/lib/stdint.h \
  ../../include/lib/kernel/stdio.h ../../tests/threads/tests.h \
  ../../include/threads/init.h ../../include/threads/thread.h \
- ../../include/lib/kernel/list.h ../../include/threads/interrupt.h
+ ../../include/lib/kernel/list.h ../../include/threads/interrupt.h \
+ ../../include/threads/synch.h
diff --git a/pintos/userprog/build/tests/threads/priority-change.o b/pintos/userprog/build/tests/threads/priority-change.o
index 599a264..284235f 100644
Binary files a/pintos/userprog/build/tests/threads/priority-change.o and b/pintos/userprog/build/tests/threads/priority-change.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-condvar.o b/pintos/userprog/build/tests/threads/priority-condvar.o
index 6b7b6c8..e4dc4e4 100644
Binary files a/pintos/userprog/build/tests/threads/priority-condvar.o and b/pintos/userprog/build/tests/threads/priority-condvar.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-chain.o b/pintos/userprog/build/tests/threads/priority-donate-chain.o
index 0167892..bf86eaa 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-chain.o and b/pintos/userprog/build/tests/threads/priority-donate-chain.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-lower.o b/pintos/userprog/build/tests/threads/priority-donate-lower.o
index 5eeeb8e..2f23743 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-lower.o and b/pintos/userprog/build/tests/threads/priority-donate-lower.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-multiple.o b/pintos/userprog/build/tests/threads/priority-donate-multiple.o
index 057b9a9..39e05c6 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-multiple.o and b/pintos/userprog/build/tests/threads/priority-donate-multiple.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-multiple2.o b/pintos/userprog/build/tests/threads/priority-donate-multiple2.o
index 3444148..b119c8f 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-multiple2.o and b/pintos/userprog/build/tests/threads/priority-donate-multiple2.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-nest.o b/pintos/userprog/build/tests/threads/priority-donate-nest.o
index 937edb7..9553aa9 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-nest.o and b/pintos/userprog/build/tests/threads/priority-donate-nest.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-one.o b/pintos/userprog/build/tests/threads/priority-donate-one.o
index 87cfca7..ee03b1e 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-one.o and b/pintos/userprog/build/tests/threads/priority-donate-one.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-donate-sema.o b/pintos/userprog/build/tests/threads/priority-donate-sema.o
index c182d4a..f0c0289 100644
Binary files a/pintos/userprog/build/tests/threads/priority-donate-sema.o and b/pintos/userprog/build/tests/threads/priority-donate-sema.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-fifo.o b/pintos/userprog/build/tests/threads/priority-fifo.o
index dfa5983..232f737 100644
Binary files a/pintos/userprog/build/tests/threads/priority-fifo.o and b/pintos/userprog/build/tests/threads/priority-fifo.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-preempt.o b/pintos/userprog/build/tests/threads/priority-preempt.o
index 7609438..ee099ad 100644
Binary files a/pintos/userprog/build/tests/threads/priority-preempt.o and b/pintos/userprog/build/tests/threads/priority-preempt.o differ
diff --git a/pintos/userprog/build/tests/threads/priority-sema.o b/pintos/userprog/build/tests/threads/priority-sema.o
index 7841851..8d15aa1 100644
Binary files a/pintos/userprog/build/tests/threads/priority-sema.o and b/pintos/userprog/build/tests/threads/priority-sema.o differ
diff --git a/pintos/userprog/build/tests/userprog/fork-once.errors b/pintos/userprog/build/tests/userprog/fork-once.errors
new file mode 100644
index 0000000..e69de29
diff --git a/pintos/userprog/build/tests/userprog/fork-once.output b/pintos/userprog/build/tests/userprog/fork-once.output
new file mode 100644
index 0000000..74eb03c
--- /dev/null
+++ b/pintos/userprog/build/tests/userprog/fork-once.output
@@ -0,0 +1,39 @@
+c[?7l[2J[0mSeaBIOS (version 1.15.0-1)
+
+
+iPXE (https://ipxe.org) 00:03.0 CA00 PCI2.10 PnP PMM+0138B4A0+012CB4A0 CA00
+Press Ctrl-B to configure iPXE (PCI 00:03.0)...                                                                               
+
+
+Booting from Hard Disk..Kernel command line: -q -f put fork-once run fork-once
+0 ~ 9fc00 1
+100000 ~ 13e0000 1
+Pintos booting with: 
+	base_mem: 0x0 ~ 0x9fc00 (Usable: 639 kB)
+	ext_mem: 0x100000 ~ 0x13e0000 (Usable: 19,328 kB)
+Calibrating timer...  419,020,800 loops/s.
+hd0: unexpected interrupt
+hd0:0: detected 329 sector (164 kB) disk, model "QEMU HARDDISK", serial "QM00001"
+hd0:1: detected 20,160 sector (9 MB) disk, model "QEMU HARDDISK", serial "QM00002"
+hd1: unexpected interrupt
+hd1:0: detected 120 sector (60 kB) disk, model "QEMU HARDDISK", serial "QM00003"
+Formatting file system...done.
+Boot complete.
+Putting 'fork-once' into the file system...
+Executing 'fork-once':
+(fork-once) begin
+(fork-once) child run
+child: exit(81)
+(fork-once) Parent: child exit status is 81
+(fork-once) end
+fork-once: exit(0)
+Execution of 'fork-once' complete.
+Timer: 476 ticks
+Thread: 425 idle ticks, 43 kernel ticks, 8 user ticks
+hd0:0: 0 reads, 0 writes
+hd0:1: 85 reads, 268 writes
+hd1:0: 120 reads, 0 writes
+Console: 976 characters output
+Keyboard: 0 keys pressed
+Exception: 0 page faults
+Powering off...
diff --git a/pintos/userprog/build/tests/userprog/fork-once.result b/pintos/userprog/build/tests/userprog/fork-once.result
new file mode 100644
index 0000000..7ef22e9
--- /dev/null
+++ b/pintos/userprog/build/tests/userprog/fork-once.result
@@ -0,0 +1 @@
+PASS
diff --git a/pintos/userprog/build/tests/userprog/open-normal.output b/pintos/userprog/build/tests/userprog/open-normal.output
index d6ed854..f3d13dd 100644
--- a/pintos/userprog/build/tests/userprog/open-normal.output
+++ b/pintos/userprog/build/tests/userprog/open-normal.output
@@ -1,19 +1,19 @@
 c[?7l[2J[0mSeaBIOS (version 1.15.0-1)
 
 
-iPXE (https://ipxe.org) 00:03.0 CA00 PCI2.10 PnP PMM+0138B4A0+012CB4A0 CA00
+iPXE (https://ipxe.org) 00:03.0 CA00 PCI2.10 PnP PMM+0FF8B4A0+0FECB4A0 CA00
 Press Ctrl-B to configure iPXE (PCI 00:03.0)...                                                                               
 
 
 Booting from Hard Disk..Kernel command line: -q -f put open-normal put sample.txt run open-normal
 0 ~ 9fc00 1
-100000 ~ 13e0000 1
+100000 ~ ffe0000 1
 Pintos booting with: 
 	base_mem: 0x0 ~ 0x9fc00 (Usable: 639 kB)
-	ext_mem: 0x100000 ~ 0x13e0000 (Usable: 19,328 kB)
+	ext_mem: 0x100000 ~ 0xffe0000 (Usable: 260,992 kB)
 Calibrating timer...  419,020,800 loops/s.
 hd0: unexpected interrupt
-hd0:0: detected 321 sector (160 kB) disk, model "QEMU HARDDISK", serial "QM00001"
+hd0:0: detected 329 sector (164 kB) disk, model "QEMU HARDDISK", serial "QM00001"
 hd0:1: detected 20,160 sector (9 MB) disk, model "QEMU HARDDISK", serial "QM00002"
 hd1: unexpected interrupt
 hd1:0: detected 121 sector (60 kB) disk, model "QEMU HARDDISK", serial "QM00003"
@@ -26,12 +26,12 @@ Executing 'open-normal':
 (open-normal) end
 open-normal: exit(0)
 Execution of 'open-normal' complete.
-Timer: 486 ticks
-Thread: 431 idle ticks, 62 kernel ticks, 2 user ticks
+Timer: 459 ticks
+Thread: 433 idle ticks, 25 kernel ticks, 1 user ticks
 hd0:0: 0 reads, 0 writes
 hd0:1: 116 reads, 280 writes
 hd1:0: 121 reads, 0 writes
-Console: 971 characters output
+Console: 972 characters output
 Keyboard: 0 keys pressed
 Exception: 0 page faults
 Powering off...
diff --git a/pintos/userprog/build/threads/init.d b/pintos/userprog/build/threads/init.d
index 9791eb3..fadf56e 100644
--- a/pintos/userprog/build/threads/init.d
+++ b/pintos/userprog/build/threads/init.d
@@ -13,10 +13,13 @@ threads/init.o: ../../threads/init.c ../../include/threads/init.h \
  ../../include/threads/mmu.h ../../include/threads/pte.h \
  ../../include/threads/vaddr.h ../../include/threads/palloc.h \
  ../../include/threads/thread.h ../../include/lib/kernel/list.h \
- ../../include/userprog/process.h ../../include/lib/kernel/hash.h \
- ../../include/lib/kernel/list.h ../../include/userprog/exception.h \
- ../../include/userprog/gdt.h ../../include/userprog/syscall.h \
+ ../../include/threads/synch.h ../../include/userprog/process.h \
+ ../../include/lib/kernel/hash.h ../../include/lib/kernel/list.h \
+ ../../include/userprog/exception.h ../../include/userprog/gdt.h \
+ ../../include/userprog/syscall.h ../../include/lib/syscall-nr.h \
+ ../../include/threads/flags.h ../../include/intrinsic.h \
+ ../../include/threads/mmu.h ../../include/filesys/file.h \
+ ../../include/filesys/off_t.h ../../include/filesys/filesys.h \
  ../../include/userprog/tss.h ../../tests/threads/tests.h \
  ../../include/devices/disk.h ../../include/lib/inttypes.h \
- ../../include/filesys/filesys.h ../../include/filesys/off_t.h \
  ../../include/filesys/fsutil.h
diff --git a/pintos/userprog/build/threads/init.o b/pintos/userprog/build/threads/init.o
index e1cbefa..94e672a 100644
Binary files a/pintos/userprog/build/threads/init.o and b/pintos/userprog/build/threads/init.o differ
diff --git a/pintos/userprog/build/threads/interrupt.d b/pintos/userprog/build/threads/interrupt.d
index 6c647dd..aeb2aec 100644
--- a/pintos/userprog/build/threads/interrupt.d
+++ b/pintos/userprog/build/threads/interrupt.d
@@ -6,8 +6,8 @@ threads/interrupt.o: ../../threads/interrupt.c \
  ../../include/lib/kernel/stdio.h ../../include/threads/flags.h \
  ../../include/threads/intr-stubs.h ../../include/threads/io.h \
  ../../include/threads/thread.h ../../include/lib/kernel/list.h \
- ../../include/threads/mmu.h ../../include/threads/pte.h \
- ../../include/threads/vaddr.h ../../include/threads/loader.h \
- ../../include/devices/timer.h ../../include/lib/round.h \
- ../../include/intrinsic.h ../../include/threads/mmu.h \
- ../../include/userprog/gdt.h
+ ../../include/threads/synch.h ../../include/threads/mmu.h \
+ ../../include/threads/pte.h ../../include/threads/vaddr.h \
+ ../../include/threads/loader.h ../../include/devices/timer.h \
+ ../../include/lib/round.h ../../include/intrinsic.h \
+ ../../include/threads/mmu.h ../../include/userprog/gdt.h
diff --git a/pintos/userprog/build/threads/interrupt.o b/pintos/userprog/build/threads/interrupt.o
index 672e3e3..bbd7528 100644
Binary files a/pintos/userprog/build/threads/interrupt.o and b/pintos/userprog/build/threads/interrupt.o differ
diff --git a/pintos/userprog/build/threads/mmu.d b/pintos/userprog/build/threads/mmu.d
index b2ee3f9..f203b21 100644
--- a/pintos/userprog/build/threads/mmu.d
+++ b/pintos/userprog/build/threads/mmu.d
@@ -5,5 +5,5 @@ threads/mmu.o: ../../threads/mmu.c ../../include/lib/stdbool.h \
  ../../include/threads/vaddr.h ../../include/threads/loader.h \
  ../../include/threads/palloc.h ../../include/threads/thread.h \
  ../../include/lib/kernel/list.h ../../include/threads/interrupt.h \
- ../../include/threads/mmu.h ../../include/intrinsic.h \
- ../../include/threads/mmu.h
+ ../../include/threads/synch.h ../../include/threads/mmu.h \
+ ../../include/intrinsic.h ../../include/threads/mmu.h
diff --git a/pintos/userprog/build/threads/synch.o b/pintos/userprog/build/threads/synch.o
index 4d9c441..e2c1477 100644
Binary files a/pintos/userprog/build/threads/synch.o and b/pintos/userprog/build/threads/synch.o differ
diff --git a/pintos/userprog/build/threads/thread.d b/pintos/userprog/build/threads/thread.d
index fee3067..bcb5d03 100644
--- a/pintos/userprog/build/threads/thread.d
+++ b/pintos/userprog/build/threads/thread.d
@@ -2,12 +2,12 @@ threads/thread.o: ../../threads/thread.c ../../include/threads/thread.h \
  ../../include/lib/debug.h ../../include/lib/kernel/list.h \
  ../../include/lib/stdbool.h ../../include/lib/stddef.h \
  ../../include/lib/stdint.h ../../include/threads/interrupt.h \
- ../../include/lib/random.h ../../include/lib/stdio.h \
- ../../include/lib/stdarg.h ../../include/lib/kernel/stdio.h \
- ../../include/lib/string.h ../../include/threads/flags.h \
- ../../include/threads/intr-stubs.h ../../include/threads/palloc.h \
- ../../include/threads/synch.h ../../include/threads/vaddr.h \
+ ../../include/threads/synch.h ../../include/lib/random.h \
+ ../../include/lib/stdio.h ../../include/lib/stdarg.h \
+ ../../include/lib/kernel/stdio.h ../../include/lib/string.h \
+ ../../include/threads/flags.h ../../include/threads/intr-stubs.h \
+ ../../include/threads/palloc.h ../../include/threads/vaddr.h \
  ../../include/threads/loader.h ../../include/intrinsic.h \
  ../../include/threads/mmu.h ../../include/threads/pte.h \
- ../../include/userprog/process.h ../../include/lib/kernel/hash.h \
- ../../include/lib/kernel/list.h
+ ../../include/threads/malloc.h ../../include/userprog/process.h \
+ ../../include/lib/kernel/hash.h ../../include/lib/kernel/list.h
diff --git a/pintos/userprog/build/threads/thread.o b/pintos/userprog/build/threads/thread.o
index b756e19..7b9018a 100644
Binary files a/pintos/userprog/build/threads/thread.o and b/pintos/userprog/build/threads/thread.o differ
diff --git a/pintos/userprog/build/userprog/exception.d b/pintos/userprog/build/userprog/exception.d
index 7242dbd..af2ed40 100644
--- a/pintos/userprog/build/userprog/exception.d
+++ b/pintos/userprog/build/userprog/exception.d
@@ -6,5 +6,6 @@ userprog/exception.o: ../../userprog/exception.c \
  ../../include/lib/kernel/stdio.h ../../include/userprog/gdt.h \
  ../../include/threads/loader.h ../../include/threads/interrupt.h \
  ../../include/threads/thread.h ../../include/lib/kernel/list.h \
- ../../include/intrinsic.h ../../include/threads/mmu.h \
- ../../include/threads/pte.h ../../include/threads/vaddr.h
+ ../../include/threads/synch.h ../../include/intrinsic.h \
+ ../../include/threads/mmu.h ../../include/threads/pte.h \
+ ../../include/threads/vaddr.h
diff --git a/pintos/userprog/build/userprog/exception.o b/pintos/userprog/build/userprog/exception.o
index 819b08f..cc3155f 100644
Binary files a/pintos/userprog/build/userprog/exception.o and b/pintos/userprog/build/userprog/exception.o differ
diff --git a/pintos/userprog/build/userprog/gdt.d b/pintos/userprog/build/userprog/gdt.d
index 992cb74..1746ca6 100644
--- a/pintos/userprog/build/userprog/gdt.d
+++ b/pintos/userprog/build/userprog/gdt.d
@@ -3,7 +3,7 @@ userprog/gdt.o: ../../userprog/gdt.c ../../include/userprog/gdt.h \
  ../../include/userprog/tss.h ../../include/lib/stdint.h \
  ../../include/threads/thread.h ../../include/lib/kernel/list.h \
  ../../include/lib/stdbool.h ../../include/lib/stddef.h \
- ../../include/threads/interrupt.h ../../include/threads/mmu.h \
- ../../include/threads/pte.h ../../include/threads/vaddr.h \
- ../../include/threads/palloc.h ../../include/intrinsic.h \
- ../../include/threads/mmu.h
+ ../../include/threads/interrupt.h ../../include/threads/synch.h \
+ ../../include/threads/mmu.h ../../include/threads/pte.h \
+ ../../include/threads/vaddr.h ../../include/threads/palloc.h \
+ ../../include/intrinsic.h ../../include/threads/mmu.h
diff --git a/pintos/userprog/build/userprog/process.d b/pintos/userprog/build/userprog/process.d
index 31b7ce0..4ae5d17 100644
--- a/pintos/userprog/build/userprog/process.d
+++ b/pintos/userprog/build/userprog/process.d
@@ -3,16 +3,17 @@ userprog/process.o: ../../userprog/process.c \
  ../../include/lib/debug.h ../../include/lib/kernel/list.h \
  ../../include/lib/stdbool.h ../../include/lib/stddef.h \
  ../../include/lib/stdint.h ../../include/threads/interrupt.h \
- ../../include/lib/kernel/hash.h ../../include/lib/kernel/list.h \
- ../../include/lib/inttypes.h ../../include/lib/round.h \
- ../../include/lib/stdio.h ../../include/lib/stdarg.h \
- ../../include/lib/kernel/stdio.h ../../include/lib/stdlib.h \
- ../../include/lib/string.h ../../include/userprog/gdt.h \
- ../../include/threads/loader.h ../../include/userprog/tss.h \
- ../../include/filesys/directory.h ../../include/devices/disk.h \
- ../../include/filesys/file.h ../../include/filesys/off_t.h \
- ../../include/filesys/filesys.h ../../include/threads/flags.h \
- ../../include/threads/init.h ../../include/threads/palloc.h \
- ../../include/threads/mmu.h ../../include/threads/pte.h \
- ../../include/threads/vaddr.h ../../include/intrinsic.h \
- ../../include/threads/mmu.h ../../include/devices/timer.h
+ ../../include/threads/synch.h ../../include/lib/kernel/hash.h \
+ ../../include/lib/kernel/list.h ../../include/lib/inttypes.h \
+ ../../include/lib/round.h ../../include/lib/stdio.h \
+ ../../include/lib/stdarg.h ../../include/lib/kernel/stdio.h \
+ ../../include/lib/stdlib.h ../../include/lib/string.h \
+ ../../include/userprog/gdt.h ../../include/threads/loader.h \
+ ../../include/userprog/tss.h ../../include/filesys/directory.h \
+ ../../include/devices/disk.h ../../include/filesys/file.h \
+ ../../include/filesys/off_t.h ../../include/filesys/filesys.h \
+ ../../include/threads/flags.h ../../include/threads/init.h \
+ ../../include/threads/palloc.h ../../include/threads/mmu.h \
+ ../../include/threads/pte.h ../../include/threads/vaddr.h \
+ ../../include/intrinsic.h ../../include/threads/mmu.h \
+ ../../include/devices/timer.h
diff --git a/pintos/userprog/build/userprog/process.o b/pintos/userprog/build/userprog/process.o
index 2b4cda8..4624cc2 100644
Binary files a/pintos/userprog/build/userprog/process.o and b/pintos/userprog/build/userprog/process.o differ
diff --git a/pintos/userprog/build/userprog/syscall.d b/pintos/userprog/build/userprog/syscall.d
index 7d6c566..256fa2d 100644
--- a/pintos/userprog/build/userprog/syscall.d
+++ b/pintos/userprog/build/userprog/syscall.d
@@ -5,10 +5,10 @@ userprog/syscall.o: ../../userprog/syscall.c \
  ../../include/lib/stdint.h ../../include/lib/kernel/stdio.h \
  ../../include/lib/syscall-nr.h ../../include/threads/interrupt.h \
  ../../include/threads/thread.h ../../include/lib/kernel/list.h \
- ../../include/threads/loader.h ../../include/userprog/gdt.h \
- ../../include/threads/flags.h ../../include/intrinsic.h \
- ../../include/threads/mmu.h ../../include/threads/pte.h \
- ../../include/threads/vaddr.h ../../include/filesys/file.h \
- ../../include/filesys/off_t.h ../../include/filesys/filesys.h \
- ../../include/userprog/process.h ../../include/lib/kernel/hash.h \
- ../../include/lib/kernel/list.h
+ ../../include/threads/synch.h ../../include/threads/loader.h \
+ ../../include/userprog/gdt.h ../../include/threads/flags.h \
+ ../../include/intrinsic.h ../../include/threads/mmu.h \
+ ../../include/threads/pte.h ../../include/threads/vaddr.h \
+ ../../include/filesys/file.h ../../include/filesys/off_t.h \
+ ../../include/filesys/filesys.h ../../include/userprog/process.h \
+ ../../include/lib/kernel/hash.h ../../include/lib/kernel/list.h
diff --git a/pintos/userprog/build/userprog/tss.d b/pintos/userprog/build/userprog/tss.d
index 3fb381a..df2f179 100644
--- a/pintos/userprog/build/userprog/tss.d
+++ b/pintos/userprog/build/userprog/tss.d
@@ -2,7 +2,8 @@ userprog/tss.o: ../../userprog/tss.c ../../include/userprog/tss.h \
  ../../include/lib/stdint.h ../../include/threads/thread.h \
  ../../include/lib/debug.h ../../include/lib/kernel/list.h \
  ../../include/lib/stdbool.h ../../include/lib/stddef.h \
- ../../include/threads/interrupt.h ../../include/userprog/gdt.h \
- ../../include/threads/loader.h ../../include/threads/palloc.h \
- ../../include/threads/vaddr.h ../../include/intrinsic.h \
- ../../include/threads/mmu.h ../../include/threads/pte.h
+ ../../include/threads/interrupt.h ../../include/threads/synch.h \
+ ../../include/userprog/gdt.h ../../include/threads/loader.h \
+ ../../include/threads/palloc.h ../../include/threads/vaddr.h \
+ ../../include/intrinsic.h ../../include/threads/mmu.h \
+ ../../include/threads/pte.h
diff --git a/pintos/userprog/process.c b/pintos/userprog/process.c
index 731f6b8..fd4ed66 100644
--- a/pintos/userprog/process.c
+++ b/pintos/userprog/process.c
@@ -19,35 +19,48 @@
 #include "threads/vaddr.h"
 #include "intrinsic.h"
 #include "devices/timer.h"
+#include "threads/synch.h"
 #ifdef VM
 #include "vm/vm.h"
 #endif
 
-static void process_cleanup (void);
-static bool load (const char *file_name, struct intr_frame *if_);
-static void initd (void *f_name);
-static void __do_fork (void *);
+struct fork_data
+{
+	struct intr_frame *_if;
+	struct thread *t;
+};
+
+
 
-/* General process initializer for initd and other process. 
+static void process_cleanup(void);
+static bool load(const char *file_name, struct intr_frame *if_);
+static void initd(void *f_name);
+static void __do_fork(void *);
+
+/* General process initializer for initd and other process.
 initd 및 기타 프로세스를 위한 일반 프로세스 초기화 함수입니다.*/
 static void
-process_init (void) {
-	struct thread *current = thread_current ();
+process_init(void)
+{
+	struct thread *current = thread_current();
 
-	current->fd_table = palloc_get_page (PAL_ZERO);
+	current->fd_table = palloc_get_page(PAL_ZERO);
 	current->fd_table->fd_node[0].type = FD_STDIN;
 	current->fd_table->fd_node[1].type = FD_STDOUT;
 }
 
-int
-process_file_open (const char *file_name) {
-	struct thread *current = thread_current ();
+int process_file_open(const char *file_name)
+{
+	struct thread *current = thread_current();
 	struct file *open_file;
 	int return_fd = -1;
 
-	for (int i = 0; i < sizeof (current->fd_table->fd_node); i++) {
-		if (current->fd_table->fd_node[i].type == FD_NONE) {
-			if ((open_file = filesys_open (file_name)) != NULL) {
+	for (int i = 0; i < sizeof(current->fd_table->fd_node); i++)
+	{
+		if (current->fd_table->fd_node[i].type == FD_NONE)
+		{
+			if ((open_file = filesys_open(file_name)) != NULL)
+			{
 				current->fd_table->fd_node[i].file = open_file;
 				current->fd_table->fd_node[i].type = FD_FILE;
 				return_fd = i;
@@ -59,64 +72,76 @@ process_file_open (const char *file_name) {
 	return return_fd;
 }
 
-int
-process_file_length (int fd) {
-	struct thread *current = thread_current ();
+int process_file_length(int fd)
+{
+	struct thread *current = thread_current();
 
-	if (fd < 0 || sizeof (current->fd_table->fd_node) <= fd) {
+	if (fd < 0 || sizeof(current->fd_table->fd_node) <= fd)
+	{
 		return -1;
 	}
-	else if (current->fd_table->fd_node[fd].type == FD_FILE) {
-		return file_length (current->fd_table->fd_node[fd].file);
+	else if (current->fd_table->fd_node[fd].type == FD_FILE)
+	{
+		return file_length(current->fd_table->fd_node[fd].file);
 	}
-	else {
+	else
+	{
 		return -1;
 	}
 }
 
-int
-process_file_read (int fd, const void *buffer, unsigned size) {
-	struct thread *current = thread_current ();
-	
-	if (fd < 0 || sizeof (current->fd_table->fd_node) <= fd) {
+int process_file_read(int fd, const void *buffer, unsigned size)
+{
+	struct thread *current = thread_current();
+
+	if (fd < 0 || sizeof(current->fd_table->fd_node) <= fd)
+	{
 		return -1;
 	}
-	else if (current->fd_table->fd_node[fd].type == FD_STDIN) {
-		return input_getc ();
+	else if (current->fd_table->fd_node[fd].type == FD_STDIN)
+	{
+		return input_getc();
 	}
-	else if (current->fd_table->fd_node[fd].type == FD_FILE) {
-		return file_read (current->fd_table->fd_node[fd].file, buffer, size);
+	else if (current->fd_table->fd_node[fd].type == FD_FILE)
+	{
+		return file_read(current->fd_table->fd_node[fd].file, buffer, size);
 	}
-	else {
+	else
+	{
 		return -1;
 	}
 }
 
-int
-process_file_write (int fd, const void *buffer, unsigned size) {
-	struct thread *current = thread_current ();
-	
-	if (fd < 0 || sizeof (current->fd_table->fd_node) <= fd) {
+int process_file_write(int fd, const void *buffer, unsigned size)
+{
+	struct thread *current = thread_current();
+
+	if (fd < 0 || sizeof(current->fd_table->fd_node) <= fd)
+	{
 		return -1;
 	}
-	else if (current->fd_table->fd_node[fd].type == FD_STDOUT) {
-		putbuf (buffer, size);
+	else if (current->fd_table->fd_node[fd].type == FD_STDOUT)
+	{
+		putbuf(buffer, size);
 		return size;
 	}
-	else if (current->fd_table->fd_node[fd].type == FD_FILE) {
-		return file_write (current->fd_table->fd_node[fd].file, buffer, size);
+	else if (current->fd_table->fd_node[fd].type == FD_FILE)
+	{
+		return file_write(current->fd_table->fd_node[fd].file, buffer, size);
 	}
-	else {
+	else
+	{
 		return -1;
 	}
 }
 
-void
-process_file_close (int fd) {
-	struct thread *current = thread_current ();
+void process_file_close(int fd)
+{
+	struct thread *current = thread_current();
 
-	if (fd < sizeof (current->fd_table->fd_node) && current->fd_table->fd_node[fd].type == FD_FILE) {
-		file_close (current->fd_table->fd_node[fd].file);
+	if (fd < sizeof(current->fd_table->fd_node) && current->fd_table->fd_node[fd].type == FD_FILE)
+	{
+		file_close(current->fd_table->fd_node[fd].file);
 		current->fd_table->fd_node[fd].type = FD_NONE;
 		current->fd_table->fd_node[fd].file = NULL;
 	}
@@ -131,17 +156,17 @@ process_file_close (int fd) {
  * 새 스레드는 process_create_initd()가 반환되기 전에 스케줄링(또는 종료)될 수 있습니다.
  * initd의 스레드 id를 반환하며, 스레드를 생성할 수 없으면 TID_ERROR를 반환합니다.
  * 반드시 한 번만 호출해야 합니다. */
-tid_t
-process_create_initd (const char *file_name) {
+tid_t process_create_initd(const char *file_name)
+{
 	char *fn_copy;
 	tid_t tid;
 
 	/* Make a copy of FILE_NAME.
 	 * Otherwise there's a race between the caller and load(). */
-	fn_copy = palloc_get_page (0);
+	fn_copy = palloc_get_page(0);
 	if (fn_copy == NULL)
 		return TID_ERROR;
-	strlcpy (fn_copy, file_name, PGSIZE);
+	strlcpy(fn_copy, file_name, PGSIZE);
 
 	/* 첫번째 인자가 파일 이름이니까 이것만 복사 */
 	char first_word[16];
@@ -150,9 +175,9 @@ process_create_initd (const char *file_name) {
 	first_word[first_len] = '\0';
 
 	/* Create a new thread to execute FILE_NAME. */
-	tid = thread_create (first_word, PRI_DEFAULT, initd, fn_copy);
+	tid = thread_create(first_word, PRI_DEFAULT, initd, fn_copy);
 	if (tid == TID_ERROR)
-		palloc_free_page (fn_copy);
+		palloc_free_page(fn_copy);
 	return tid;
 }
 
@@ -160,27 +185,34 @@ process_create_initd (const char *file_name) {
 /* 첫 번째 사용자 프로세스를 실행하는 스레드 함수입니다. */
 
 static void
-initd (void *f_name) {
+initd(void *f_name)
+{
 #ifdef VM
-	supplemental_page_table_init (&thread_current ()->spt);
+	supplemental_page_table_init(&thread_current()->spt);
 #endif
 
-	process_init ();
+	process_init();
 
-	if (process_exec (f_name) < 0)
+	if (process_exec(f_name) < 0)
 		PANIC("Fail to launch initd\n");
-	NOT_REACHED ();
+	NOT_REACHED();
 }
 
 /* Clones the current process as `name`. Returns the new process's thread id, or
  * TID_ERROR if the thread cannot be created. */
 /* 현재 프로세스를 `name`으로 복제합니다. 새 프로세스의 스레드 id를 반환하며,
  * 스레드를 생성할 수 없으면 TID_ERROR를 반환합니다. */
-tid_t
-process_fork (const char *name, struct intr_frame *if_ UNUSED) {
+tid_t process_fork(const char *name, struct intr_frame *if_ UNUSED)
+{
+	struct fork_data *fdata = malloc(sizeof(struct fork_data));
+	fdata->t = thread_current();
+	fdata->_if = if_;
 	/* Clone current thread to new thread.*/
-	return thread_create (name,
-			PRI_DEFAULT, __do_fork, thread_current ());
+	tid_t child_tid = thread_create(name, PRI_DEFAULT, __do_fork, fdata);
+
+	sema_down(&fdata->t->self_jail);
+
+	return child_tid;
 }
 
 #ifndef VM
@@ -189,40 +221,54 @@ process_fork (const char *name, struct intr_frame *if_ UNUSED) {
 /* 이 함수를 pml4_for_each에 전달하여 부모의 주소 공간을 복제합니다.
  * 이 코드는 project 2에서만 사용됩니다. */
 static bool
-duplicate_pte (uint64_t *pte, void *va, void *aux) {
-	struct thread *current = thread_current ();
-	struct thread *parent = (struct thread *) aux;
+duplicate_pte(uint64_t *pte, void *va, void *aux)
+{
+	struct thread *current = thread_current();
+	struct thread *parent = (struct thread *)aux;
 	void *parent_page;
 	void *newpage;
 	bool writable;
 
 	/* 1. TODO: If the parent_page is kernel page, then return immediately. */
 	/* 1. TODO: parent_page가 커널 페이지라면 즉시 반환합니다. */
-
+	if (is_kernel_vaddr(va))
+	{
+		return true;
+	}
 	/* 2. Resolve VA from the parent's page map level 4. */
 	/* 2. 부모의 pml4에서 VA를 해석합니다. */
 
-	parent_page = pml4_get_page (parent->pml4, va);
+	parent_page = pml4_get_page(parent->pml4, va);
+	if (parent_page == NULL)
+	{
+		return false;
+	}
 
+	writable = is_writable(pte);
 	/* 3. TODO: Allocate new PAL_USER page for the child and set result to
 	 *    TODO: NEWPAGE. */
 	/* 3. TODO: 자식 프로세스를 위해 PAL_USER 페이지를 새로 할당하고, 결과를 NEWPAGE에 저장합니다. */
-
-
+	newpage = palloc_get_page(PAL_USER | PAL_ZERO);
+	if (newpage == NULL)
+	{
+		return false;
+	}
 	/* 4. TODO: Duplicate parent's page to the new page and
 	 *    TODO: check whether parent's page is writable or not (set WRITABLE
 	 *    TODO: according to the result). */
 	/* 4. TODO: 부모의 페이지를 새 페이지에 복제하고,
- *    TODO: 부모의 페이지가 쓰기 가능한지 확인하여 WRITABLE 값을 설정합니다. */
-
+	 *    TODO: 부모의 페이지가 쓰기 가능한지 확인하여 WRITABLE 값을 설정합니다. */
+	memcpy(newpage, parent_page, PGSIZE);
 	/* 5. Add new page to child's page table at address VA with WRITABLE
 	 *    permission. */
 	/* 5. WRITABLE 권한으로 자식의 페이지 테이블에 VA 주소에 새 페이지를 추가합니다. */
 
-	if (!pml4_set_page (current->pml4, va, newpage, writable)) {
+	if (!pml4_set_page(current->pml4, va, newpage, writable))
+	{
+		palloc_free_page(newpage);
+		return false;
 		/* 6. TODO: if fail to insert page, do error handling. */
 		/* 6. TODO: 페이지 삽입에 실패하면 에러 처리를 합니다. */
-
 	}
 	return true;
 }
@@ -237,51 +283,56 @@ duplicate_pte (uint64_t *pte, void *va, void *aux) {
  *       즉, process_fork의 두 번째 인자를 이 함수에 전달해야 합니다. */
 
 static void
-__do_fork (void *aux) {
+__do_fork(void *aux)
+{
 	struct intr_frame if_;
-	struct thread *parent = (struct thread *) aux;
-	struct thread *current = thread_current ();
+	struct fork_data *fdata = (struct fork_data *)aux;
+	struct thread *parent = fdata->t;
+	struct thread *current = thread_current();
 	/* TODO: somehow pass the parent_if. (i.e. process_fork()'s if_) */
-	struct intr_frame *parent_if;
+	struct intr_frame *parent_if = fdata->_if;
 	bool succ = true;
 
+	current->parent = parent;
+
 	/* 1. Read the cpu context to local stack. */
-	memcpy (&if_, parent_if, sizeof (struct intr_frame));
+	memcpy(&if_, parent_if, sizeof(struct intr_frame));
 
 	/* 2. Duplicate PT */
 	current->pml4 = pml4_create();
 	if (current->pml4 == NULL)
 		goto error;
 
-	process_activate (current);
+	process_activate(current);
 #ifdef VM
-	supplemental_page_table_init (&current->spt);
-	if (!supplemental_page_table_copy (&current->spt, &parent->spt))
+	supplemental_page_table_init(&current->spt);
+	if (!supplemental_page_table_copy(&current->spt, &parent->spt))
 		goto error;
 #else
-	if (!pml4_for_each (parent->pml4, duplicate_pte, parent))
+	if (!pml4_for_each(parent->pml4, duplicate_pte, current->parent))
 		goto error;
 #endif
 
 	/* TODO: Your code goes here.
-	 * TODO: Hint) To duplicate the file object, use `file_duplicate`
-	 * TODO:       in include/filesys/file.h. Note that parent should not return
-	 * TODO:       from the fork() until this function successfully duplicates
-	 * TODO:       the resources of parent.*/
+	  TODO: Hint) To duplicate the file object, use `file_duplicate`
+	  TODO:       in include/filesys/file.h. Note that parent should not return
+	  TODO:       from the fork() until this function successfully duplicates
+	  TODO:       the resources of parent.*/
 
-	process_init ();
+	process_init();
+	if_.R.rax = 0;
 
 	/* Finally, switch to the newly created process. */
 	if (succ)
-		do_iret (&if_);
+		do_iret(&if_);
 error:
-	thread_exit ();
+	thread_exit();
 }
 
 /* Switch the current execution context to the f_name.
  * Returns -1 on fail. */
-int
-process_exec (void *f_name) {
+int process_exec(void *f_name)
+{
 	char *file_name = f_name;
 	bool success;
 
@@ -294,22 +345,21 @@ process_exec (void *f_name) {
 	_if.eflags = FLAG_IF | FLAG_MBS;
 
 	/* We first kill the current context */
-	process_cleanup ();
+	process_cleanup();
 
 	/* And then load the binary */
-	success = load (file_name, &_if);
+	success = load(file_name, &_if);
 
 	/* If load failed, quit. */
-	palloc_free_page (file_name);
+	palloc_free_page(file_name);
 	if (!success)
 		return -1;
 
 	/* Start switched process. */
-	do_iret (&_if);
-	NOT_REACHED ();
+	do_iret(&_if);
+	NOT_REACHED();
 }
 
-
 /* Waits for thread TID to die and returns its exit status.  If
  * it was terminated by the kernel (i.e. killed due to an
  * exception), returns -1.  If TID is invalid or if it was not a
@@ -318,55 +368,79 @@ process_exec (void *f_name) {
  * immediately, without waiting.
  *
  * This function will be implemented in problem 2-2.  For now, it
- * does nothing. 
- * 
+ * does nothing.
+ *
  * 스레드 TID가 종료될 때까지 기다렸다가, 그 종료 상태(exit status)를 반환합니다.
 만약 그 스레드가 커널에 의해 종료되었으면(예: 예외 때문에 강제 종료된 경우), -1을 반환합니다.
-TID가 유효하지 않거나, 호출한 프로세스의 자식이 아니거나, 
-해당 TID에 대해 process_wait()가 이미 성공적으로 호출된 적이 있다면, 
+TID가 유효하지 않거나, 호출한 프로세스의 자식이 아니거나,
+해당 TID에 대해 process_wait()가 이미 성공적으로 호출된 적이 있다면,
 기다리지 않고 즉시 -1을 반환합니다.
-이 함수는 문제 2-2에서 구현될 예정입니다. 
+이 함수는 문제 2-2에서 구현될 예정입니다.
 현재는 아무 동작도 하지 않습니다.*/
-int
-process_wait (tid_t child_tid UNUSED) {
+int process_wait(tid_t child_tid UNUSED)
+{
 	/* XXX: Hint) The pintos exit if process_wait (initd), we recommend you
 	 * XXX:       to add infinite loop here before
 	 * XXX:       implementing the process_wait. */
 
-	 /* XXX: 힌트) PintOS는 process_wait(initd)가 호출되면 종료됩니다.
+	/* XXX: 힌트) PintOS는 process_wait(initd)가 호출되면 종료됩니다.
 	 * XXX: process_wait를 구현하기 전에 이 부분에 무한 루프를 추가하는 것을 권장합니다. */
+	struct thread *curr = thread_current();
+	struct list_elem *elem;
+	if (!list_empty(&curr->child_dying_list))
+	{
+		for (elem = list_front(&curr->child_dying_list);
+			 elem != list_end(&curr->child_dying_list);
+			 elem = list_next(elem))
+		{
+			struct dying_msg *dmsg = list_entry(elem, struct dying_msg, elem);
+			if (dmsg->tid == child_tid)
+			{
+				int result = dmsg->dying_msg;
+				list_remove(elem);
+				free(dmsg);
+				return result;
+			}
+		}
+	}
 
 	//  while(true){}
-	timer_sleep(400);
+	// timer_sleep(400);
 	return -1;
 }
 
 /* Exit the process. This function is called by thread_exit (). */
-void
-process_exit (void) {
-	struct thread *curr = thread_current ();
+void process_exit(void)
+{
+	struct thread *curr = thread_current();
 	/* TODO: Your code goes here.
 	 * TODO: Implement process termination message (see
 	 * TODO: project2/process_termination.html).
 	 * TODO: We recommend you to implement process resource cleanup here. */
+	
+	
+	list_push_back(&curr->parent->child_dying_list, &curr->dmsg->elem);
+	sema_up(&curr->parent->self_jail);
 
-	process_cleanup ();
+	process_cleanup();
 }
 
 /* Free the current process's resources. */
 static void
-process_cleanup (void) {
-	struct thread *curr = thread_current ();
+process_cleanup(void)
+{
+	struct thread *curr = thread_current();
 
 #ifdef VM
-	supplemental_page_table_kill (&curr->spt);
+	supplemental_page_table_kill(&curr->spt);
 #endif
 
 	uint64_t *pml4;
 	/* Destroy the current process's page directory and switch back
 	 * to the kernel-only page directory. */
 	pml4 = curr->pml4;
-	if (pml4 != NULL) {
+	if (pml4 != NULL)
+	{
 		/* Correct ordering here is crucial.  We must set
 		 * cur->pagedir to NULL before switching page directories,
 		 * so that a timer interrupt can't switch back to the
@@ -375,20 +449,20 @@ process_cleanup (void) {
 		 * directory, or our active page directory will be one
 		 * that's been freed (and cleared). */
 		curr->pml4 = NULL;
-		pml4_activate (NULL);
-		pml4_destroy (pml4);
+		pml4_activate(NULL);
+		pml4_destroy(pml4);
 	}
 }
 
 /* Sets up the CPU for running user code in the nest thread.
  * This function is called on every context switch. */
-void
-process_activate (struct thread *next) {
+void process_activate(struct thread *next)
+{
 	/* Activate thread's page tables. */
-	pml4_activate (next->pml4);
+	pml4_activate(next->pml4);
 
 	/* Set thread's kernel stack for use in processing interrupts. */
-	tss_update (next);
+	tss_update(next);
 }
 
 /* We load ELF binaries.  The following definitions are taken
@@ -397,22 +471,23 @@ process_activate (struct thread *next) {
 /* ELF types.  See [ELF1] 1-2. */
 #define EI_NIDENT 16
 
-#define PT_NULL    0            /* Ignore. */
-#define PT_LOAD    1            /* Loadable segment. */
-#define PT_DYNAMIC 2            /* Dynamic linking info. */
-#define PT_INTERP  3            /* Name of dynamic loader. */
-#define PT_NOTE    4            /* Auxiliary info. */
-#define PT_SHLIB   5            /* Reserved. */
-#define PT_PHDR    6            /* Program header table. */
-#define PT_STACK   0x6474e551   /* Stack segment. */
+#define PT_NULL 0			/* Ignore. */
+#define PT_LOAD 1			/* Loadable segment. */
+#define PT_DYNAMIC 2		/* Dynamic linking info. */
+#define PT_INTERP 3			/* Name of dynamic loader. */
+#define PT_NOTE 4			/* Auxiliary info. */
+#define PT_SHLIB 5			/* Reserved. */
+#define PT_PHDR 6			/* Program header table. */
+#define PT_STACK 0x6474e551 /* Stack segment. */
 
-#define PF_X 1          /* Executable. */
-#define PF_W 2          /* Writable. */
-#define PF_R 4          /* Readable. */
+#define PF_X 1 /* Executable. */
+#define PF_W 2 /* Writable. */
+#define PF_R 4 /* Readable. */
 
 /* Executable header.  See [ELF1] 1-4 to 1-8.
  * This appears at the very beginning of an ELF binary. */
-struct ELF64_hdr {
+struct ELF64_hdr
+{
 	unsigned char e_ident[EI_NIDENT];
 	uint16_t e_type;
 	uint16_t e_machine;
@@ -429,7 +504,8 @@ struct ELF64_hdr {
 	uint16_t e_shstrndx;
 };
 
-struct ELF64_PHDR {
+struct ELF64_PHDR
+{
 	uint32_t p_type;
 	uint32_t p_flags;
 	uint64_t p_offset;
@@ -444,19 +520,20 @@ struct ELF64_PHDR {
 #define ELF ELF64_hdr
 #define Phdr ELF64_PHDR
 
-static bool setup_stack (struct intr_frame *if_);
-static bool validate_segment (const struct Phdr *, struct file *);
-static bool load_segment (struct file *file, off_t ofs, uint8_t *upage,
-		uint32_t read_bytes, uint32_t zero_bytes,
-		bool writable);
+static bool setup_stack(struct intr_frame *if_);
+static bool validate_segment(const struct Phdr *, struct file *);
+static bool load_segment(struct file *file, off_t ofs, uint8_t *upage,
+						 uint32_t read_bytes, uint32_t zero_bytes,
+						 bool writable);
 
 /* Loads an ELF executable from FILE_NAME into the current thread.
  * Stores the executable's entry point into *RIP
  * and its initial stack pointer into *RSP.
  * Returns true if successful, false otherwise. */
 static bool
-load (const char *file_name, struct intr_frame *if_) {
-	struct thread *t = thread_current ();
+load(const char *file_name, struct intr_frame *if_)
+{
+	struct thread *t = thread_current();
 	struct ELF ehdr;
 	struct file *file = NULL;
 	off_t file_ofs;
@@ -464,85 +541,87 @@ load (const char *file_name, struct intr_frame *if_) {
 	int i;
 
 	/* Allocate and activate page directory. */
-	t->pml4 = pml4_create ();
+	t->pml4 = pml4_create();
 	if (t->pml4 == NULL)
 		goto done;
-	process_activate (thread_current ());
+	process_activate(thread_current());
 
 	/* Open executable file. */
-	file = filesys_open (thread_name ());
-	if (file == NULL) {
-		printf ("load: %s: open failed\n", thread_name ());
+	file = filesys_open(thread_name());
+	if (file == NULL)
+	{
+		printf("load: %s: open failed\n", thread_name());
 		goto done;
 	}
 
 	/* Read and verify executable header. */
-	if (file_read (file, &ehdr, sizeof ehdr) != sizeof ehdr
-			|| memcmp (ehdr.e_ident, "\177ELF\2\1\1", 7)
-			|| ehdr.e_type != 2
-			|| ehdr.e_machine != 0x3E // amd64
-			|| ehdr.e_version != 1
-			|| ehdr.e_phentsize != sizeof (struct Phdr)
-			|| ehdr.e_phnum > 1024) {
-		printf ("load: %s: error loading executable\n", thread_name ());
+	if (file_read(file, &ehdr, sizeof ehdr) != sizeof ehdr || memcmp(ehdr.e_ident, "\177ELF\2\1\1", 7) || ehdr.e_type != 2 || ehdr.e_machine != 0x3E // amd64
+		|| ehdr.e_version != 1 || ehdr.e_phentsize != sizeof(struct Phdr) || ehdr.e_phnum > 1024)
+	{
+		printf("load: %s: error loading executable\n", thread_name());
 		goto done;
 	}
 
 	/* Read program headers. */
 	file_ofs = ehdr.e_phoff;
-	for (i = 0; i < ehdr.e_phnum; i++) {
+	for (i = 0; i < ehdr.e_phnum; i++)
+	{
 		struct Phdr phdr;
 
-		if (file_ofs < 0 || file_ofs > file_length (file))
+		if (file_ofs < 0 || file_ofs > file_length(file))
 			goto done;
-		file_seek (file, file_ofs);
+		file_seek(file, file_ofs);
 
-		if (file_read (file, &phdr, sizeof phdr) != sizeof phdr)
+		if (file_read(file, &phdr, sizeof phdr) != sizeof phdr)
 			goto done;
 		file_ofs += sizeof phdr;
-		switch (phdr.p_type) {
-			case PT_NULL:
-			case PT_NOTE:
-			case PT_PHDR:
-			case PT_STACK:
-			default:
-				/* Ignore this segment. */
-				break;
-			case PT_DYNAMIC:
-			case PT_INTERP:
-			case PT_SHLIB:
-				goto done;
-			case PT_LOAD:
-				if (validate_segment (&phdr, file)) {
-					bool writable = (phdr.p_flags & PF_W) != 0;
-					uint64_t file_page = phdr.p_offset & ~PGMASK;
-					uint64_t mem_page = phdr.p_vaddr & ~PGMASK;
-					uint64_t page_offset = phdr.p_vaddr & PGMASK;
-					uint32_t read_bytes, zero_bytes;
-					if (phdr.p_filesz > 0) {
-						/* Normal segment.
-						 * Read initial part from disk and zero the rest. */
-						read_bytes = page_offset + phdr.p_filesz;
-						zero_bytes = (ROUND_UP (page_offset + phdr.p_memsz, PGSIZE)
-								- read_bytes);
-					} else {
-						/* Entirely zero.
-						 * Don't read anything from disk. */
-						read_bytes = 0;
-						zero_bytes = ROUND_UP (page_offset + phdr.p_memsz, PGSIZE);
-					}
-					if (!load_segment (file, file_page, (void *) mem_page,
-								read_bytes, zero_bytes, writable))
-						goto done;
+		switch (phdr.p_type)
+		{
+		case PT_NULL:
+		case PT_NOTE:
+		case PT_PHDR:
+		case PT_STACK:
+		default:
+			/* Ignore this segment. */
+			break;
+		case PT_DYNAMIC:
+		case PT_INTERP:
+		case PT_SHLIB:
+			goto done;
+		case PT_LOAD:
+			if (validate_segment(&phdr, file))
+			{
+				bool writable = (phdr.p_flags & PF_W) != 0;
+				uint64_t file_page = phdr.p_offset & ~PGMASK;
+				uint64_t mem_page = phdr.p_vaddr & ~PGMASK;
+				uint64_t page_offset = phdr.p_vaddr & PGMASK;
+				uint32_t read_bytes, zero_bytes;
+				if (phdr.p_filesz > 0)
+				{
+					/* Normal segment.
+					 * Read initial part from disk and zero the rest. */
+					read_bytes = page_offset + phdr.p_filesz;
+					zero_bytes = (ROUND_UP(page_offset + phdr.p_memsz, PGSIZE) - read_bytes);
 				}
 				else
+				{
+					/* Entirely zero.
+					 * Don't read anything from disk. */
+					read_bytes = 0;
+					zero_bytes = ROUND_UP(page_offset + phdr.p_memsz, PGSIZE);
+				}
+				if (!load_segment(file, file_page, (void *)mem_page,
+								  read_bytes, zero_bytes, writable))
 					goto done;
-				break;
+			}
+			else
+				goto done;
+			break;
 		}
 	}
 
 	/* Set up stack. */
-	if (!setup_stack (if_))
+	if (!setup_stack(if_))
 		goto done;
 
 	/* Start address. */
@@ -552,33 +631,36 @@ load (const char *file_name, struct intr_frame *if_) {
 	char *token, *save, *cmd_str[64];
 	int count = 0;
 	token = strtok_r(file_name, " ", &save);
-	do {
+	do
+	{
 		cmd_str[count++] = token;
 		token = strtok_r(NULL, " ", &save);
 	} while (token != NULL);
 
 	/* 스택의 rsp부터 차근차근 반대로 삽입 */
 	char *cmd_ptr[64];
-	for (int i = count - 1; i >= 0; --i) {
-		if_->rsp = ((char*)if_->rsp - (strlen(cmd_str[i]) + 1));
+	for (int i = count - 1; i >= 0; --i)
+	{
+		if_->rsp = ((char *)if_->rsp - (strlen(cmd_str[i]) + 1));
 		memcpy(if_->rsp, cmd_str[i], (strlen(cmd_str[i]) + 1));
 		cmd_ptr[i] = if_->rsp;
 	}
 
 	/* 인자 문자열을 넣은 후 패딩 */
 	uintptr_t padding = if_->rsp % 8;
-	if_->rsp = ((char*)if_->rsp - (padding + 8));	// 패딩 뿐만 아니라 문자열 주소의 끝을 나타내는 NULL을 위해서 8 추가
-	memset (if_->rsp, 0, (padding + 8));
+	if_->rsp = ((char *)if_->rsp - (padding + 8)); // 패딩 뿐만 아니라 문자열 주소의 끝을 나타내는 NULL을 위해서 8 추가
+	memset(if_->rsp, 0, (padding + 8));
 
 	/* 인자 문자열의 주소를 삽입 */
-	for (int i = count - 1; i >= 0; --i) {
-		if_->rsp = ((char*)if_->rsp - (8));
+	for (int i = count - 1; i >= 0; --i)
+	{
+		if_->rsp = ((char *)if_->rsp - (8));
 		memcpy(if_->rsp, &cmd_ptr[i], (8));
 	}
 
 	/* 가짜 return address 삽입 */
-	if_->rsp = ((char*)if_->rsp - (8));
-	memset (if_->rsp, 0, (8));
+	if_->rsp = ((char *)if_->rsp - (8));
+	memset(if_->rsp, 0, (8));
 
 	/* rdi, rsi 레지스터 삽입 */
 	if_->R.rdi = count;
@@ -588,21 +670,21 @@ load (const char *file_name, struct intr_frame *if_) {
 
 done:
 	/* We arrive here whether the load is successful or not. */
-	file_close (file);
+	file_close(file);
 	return success;
 }
 
-
 /* Checks whether PHDR describes a valid, loadable segment in
  * FILE and returns true if so, false otherwise. */
 static bool
-validate_segment (const struct Phdr *phdr, struct file *file) {
+validate_segment(const struct Phdr *phdr, struct file *file)
+{
 	/* p_offset and p_vaddr must have the same page offset. */
 	if ((phdr->p_offset & PGMASK) != (phdr->p_vaddr & PGMASK))
 		return false;
 
 	/* p_offset must point within FILE. */
-	if (phdr->p_offset > (uint64_t) file_length (file))
+	if (phdr->p_offset > (uint64_t)file_length(file))
 		return false;
 
 	/* p_memsz must be at least as big as p_filesz. */
@@ -615,9 +697,9 @@ validate_segment (const struct Phdr *phdr, struct file *file) {
 
 	/* The virtual memory region must both start and end within the
 	   user address space range. */
-	if (!is_user_vaddr ((void *) phdr->p_vaddr))
+	if (!is_user_vaddr((void *)phdr->p_vaddr))
 		return false;
-	if (!is_user_vaddr ((void *) (phdr->p_vaddr + phdr->p_memsz)))
+	if (!is_user_vaddr((void *)(phdr->p_vaddr + phdr->p_memsz)))
 		return false;
 
 	/* The region cannot "wrap around" across the kernel virtual
@@ -643,7 +725,7 @@ validate_segment (const struct Phdr *phdr, struct file *file) {
  * outside of #ifndef macro. */
 
 /* load() helpers. */
-static bool install_page (void *upage, void *kpage, bool writable);
+static bool install_page(void *upage, void *kpage, bool writable);
 
 /* Loads a segment starting at offset OFS in FILE at address
  * UPAGE.  In total, READ_BYTES + ZERO_BYTES bytes of virtual
@@ -660,14 +742,16 @@ static bool install_page (void *upage, void *kpage, bool writable);
  * Return true if successful, false if a memory allocation error
  * or disk read error occurs. */
 static bool
-load_segment (struct file *file, off_t ofs, uint8_t *upage,
-		uint32_t read_bytes, uint32_t zero_bytes, bool writable) {
-	ASSERT ((read_bytes + zero_bytes) % PGSIZE == 0);
-	ASSERT (pg_ofs (upage) == 0);
-	ASSERT (ofs % PGSIZE == 0);
-
-	file_seek (file, ofs);
-	while (read_bytes > 0 || zero_bytes > 0) {
+load_segment(struct file *file, off_t ofs, uint8_t *upage,
+			 uint32_t read_bytes, uint32_t zero_bytes, bool writable)
+{
+	ASSERT((read_bytes + zero_bytes) % PGSIZE == 0);
+	ASSERT(pg_ofs(upage) == 0);
+	ASSERT(ofs % PGSIZE == 0);
+
+	file_seek(file, ofs);
+	while (read_bytes > 0 || zero_bytes > 0)
+	{
 		/* Do calculate how to fill this page.
 		 * We will read PAGE_READ_BYTES bytes from FILE
 		 * and zero the final PAGE_ZERO_BYTES bytes. */
@@ -675,21 +759,23 @@ load_segment (struct file *file, off_t ofs, uint8_t *upage,
 		size_t page_zero_bytes = PGSIZE - page_read_bytes;
 
 		/* Get a page of memory. */
-		uint8_t *kpage = palloc_get_page (PAL_USER);
+		uint8_t *kpage = palloc_get_page(PAL_USER);
 		if (kpage == NULL)
 			return false;
 
 		/* Load this page. */
-		if (file_read (file, kpage, page_read_bytes) != (int) page_read_bytes) {
-			palloc_free_page (kpage);
+		if (file_read(file, kpage, page_read_bytes) != (int)page_read_bytes)
+		{
+			palloc_free_page(kpage);
 			return false;
 		}
-		memset (kpage + page_read_bytes, 0, page_zero_bytes);
+		memset(kpage + page_read_bytes, 0, page_zero_bytes);
 
 		/* Add the page to the process's address space. */
-		if (!install_page (upage, kpage, writable)) {
+		if (!install_page(upage, kpage, writable))
+		{
 			printf("fail\n");
-			palloc_free_page (kpage);
+			palloc_free_page(kpage);
 			return false;
 		}
 
@@ -703,17 +789,19 @@ load_segment (struct file *file, off_t ofs, uint8_t *upage,
 
 /* Create a minimal stack by mapping a zeroed page at the USER_STACK */
 static bool
-setup_stack (struct intr_frame *if_) {
+setup_stack(struct intr_frame *if_)
+{
 	uint8_t *kpage;
 	bool success = false;
 
-	kpage = palloc_get_page (PAL_USER | PAL_ZERO);
-	if (kpage != NULL) {
-		success = install_page (((uint8_t *) USER_STACK) - PGSIZE, kpage, true);
+	kpage = palloc_get_page(PAL_USER | PAL_ZERO);
+	if (kpage != NULL)
+	{
+		success = install_page(((uint8_t *)USER_STACK) - PGSIZE, kpage, true);
 		if (success)
 			if_->rsp = USER_STACK;
 		else
-			palloc_free_page (kpage);
+			palloc_free_page(kpage);
 	}
 	return success;
 }
@@ -728,13 +816,13 @@ setup_stack (struct intr_frame *if_) {
  * Returns true on success, false if UPAGE is already mapped or
  * if memory allocation fails. */
 static bool
-install_page (void *upage, void *kpage, bool writable) {
-	struct thread *t = thread_current ();
+install_page(void *upage, void *kpage, bool writable)
+{
+	struct thread *t = thread_current();
 
 	/* Verify that there's not already a page at that virtual
 	 * address, then map our page there. */
-	return (pml4_get_page (t->pml4, upage) == NULL
-			&& pml4_set_page (t->pml4, upage, kpage, writable));
+	return (pml4_get_page(t->pml4, upage) == NULL && pml4_set_page(t->pml4, upage, kpage, writable));
 }
 #else
 /* From here, codes will be used after project 3.
@@ -742,7 +830,8 @@ install_page (void *upage, void *kpage, bool writable) {
  * upper block. */
 
 static bool
-lazy_load_segment (struct page *page, void *aux) {
+lazy_load_segment(struct page *page, void *aux)
+{
 	/* TODO: Load the segment from the file */
 	/* TODO: This called when the first page fault occurs on address VA. */
 	/* TODO: VA is available when calling this function. */
@@ -763,13 +852,15 @@ lazy_load_segment (struct page *page, void *aux) {
  * Return true if successful, false if a memory allocation error
  * or disk read error occurs. */
 static bool
-load_segment (struct file *file, off_t ofs, uint8_t *upage,
-		uint32_t read_bytes, uint32_t zero_bytes, bool writable) {
-	ASSERT ((read_bytes + zero_bytes) % PGSIZE == 0);
-	ASSERT (pg_ofs (upage) == 0);
-	ASSERT (ofs % PGSIZE == 0);
-
-	while (read_bytes > 0 || zero_bytes > 0) {
+load_segment(struct file *file, off_t ofs, uint8_t *upage,
+			 uint32_t read_bytes, uint32_t zero_bytes, bool writable)
+{
+	ASSERT((read_bytes + zero_bytes) % PGSIZE == 0);
+	ASSERT(pg_ofs(upage) == 0);
+	ASSERT(ofs % PGSIZE == 0);
+
+	while (read_bytes > 0 || zero_bytes > 0)
+	{
 		/* Do calculate how to fill this page.
 		 * We will read PAGE_READ_BYTES bytes from FILE
 		 * and zero the final PAGE_ZERO_BYTES bytes. */
@@ -778,8 +869,8 @@ load_segment (struct file *file, off_t ofs, uint8_t *upage,
 
 		/* TODO: Set up aux to pass information to the lazy_load_segment. */
 		void *aux = NULL;
-		if (!vm_alloc_page_with_initializer (VM_ANON, upage,
-					writable, lazy_load_segment, aux))
+		if (!vm_alloc_page_with_initializer(VM_ANON, upage,
+											writable, lazy_load_segment, aux))
 			return false;
 
 		/* Advance. */
@@ -792,18 +883,28 @@ load_segment (struct file *file, off_t ofs, uint8_t *upage,
 
 /* Create a PAGE of stack at the USER_STACK. Return true on success. */
 static bool
-setup_stack (struct intr_frame *if_) {
+setup_stack(struct intr_frame *if_)
+{
 	bool success = false;
-	void *stack_bottom = (void *) (((uint8_t *) USER_STACK) - PGSIZE);
+	void *stack_bottom = (void *)(((uint8_t *)USER_STACK) - PGSIZE);
 
 	/* TODO: Map the stack on stack_bottom and claim the page immediately.
 	스택의 바닥(stack_bottom)에 스택을 매핑하고, 즉시 해당 페이지를 할당(claim)하세요.
 	 * TODO: If success, set the rsp accordingly.
 	 성공했다면, rsp 레지스터를 그에 맞게 설정하세요
-	 * TODO: You should mark the page is stack. 
+	 * TODO: You should mark the page is stack.
 	 해당 페이지를 스택으로 표시*/
 	/* TODO: Your code goes here */
 
 	return success;
 }
 #endif /* VM */
+
+// struct dying_msg *dying_msg_init(tid_t tid)
+// {
+// 	struct dying_msg *dmsg = (struct dying_msg *)malloc(sizeof(struct dying_msg));
+// 	dmsg->tid = tid;
+// 	dmsg->dying_msg = 0;
+
+// 	return dmsg;
+// }
\ No newline at end of file
diff --git a/pintos/userprog/syscall.c b/pintos/userprog/syscall.c
index 8ae5ca6..a602298 100644
--- a/pintos/userprog/syscall.c
+++ b/pintos/userprog/syscall.c
@@ -1,15 +1,5 @@
 #include "userprog/syscall.h"
-#include <stdio.h>
-#include <syscall-nr.h>
-#include "threads/interrupt.h"
-#include "threads/thread.h"
-#include "threads/loader.h"
-#include "userprog/gdt.h"
-#include "threads/flags.h"
-#include "intrinsic.h"
-#include "filesys/file.h"
-#include "filesys/filesys.h"
-#include "userprog/process.h"
+
 
 void syscall_entry (void);
 void syscall_handler (struct intr_frame *);
@@ -62,10 +52,12 @@ syscall_handler (struct intr_frame *f UNUSED) {
 			sys_exit (f->R.rdi);
 			break;
 		case SYS_FORK:
+			f->R.rax = sys_fork(f->R.rdi,f);
 			break;
 		case SYS_EXEC:
 			break;
 		case SYS_WAIT:
+			f->R.rax = sys_wait(f->R.rdi);
 			break;
 		case SYS_CREATE:
 			f->R.rax = sys_create (f-> R.rdi, f -> R.rsi);
@@ -104,14 +96,22 @@ void
 sys_halt (void) {
 	power_off (); 
 }
-
+int sys_wait (tid_t tid){
+	process_wait(tid);
+}
 /* exit로 호출한 스레드를 종료하는 함수 */
 void
 sys_exit (int status) {
 	printf ("%s: exit(%d)\n", thread_name (), status);
+	thread_current()->dmsg->dying_msg = status;
 	thread_exit ();
 }
 
+tid_t sys_fork (const char *thread_name, struct intr_frame * _if){
+
+	return process_fork(thread_name, _if);
+}
+
 /* filesys_create를 호출하며 새로운 파일을 만듭니다. */
 bool
 sys_create (const char *file, unsigned initial_size) {
